// ngx-joyride/esm2015/lib/components/close-button/close-button.component.js
import { Component } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export class JoyrideCloseButtonComponent {
}
JoyrideCloseButtonComponent.ɵfac = function JoyrideCloseButtonComponent_Factory(t) { return new (t || JoyrideCloseButtonComponent)(); };
JoyrideCloseButtonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: JoyrideCloseButtonComponent, selectors: [["joy-close-button"]], decls: 3, vars: 0, consts: [["viewBox", "0 0 25 25", "xmlns", "http://www.w3.org/2000/svg"], ["x1", "1", "y1", "24", "x2", "24", "y2", "1", "stroke", "black", "stroke-width", "3"], ["x1", "1", "y1", "1", "x2", "24", "y2", "24", "stroke", "black", "stroke-width", "3"]], template: function JoyrideCloseButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵnamespaceSVG();
        ɵngcc0.ɵɵelementStart(0, "svg", 0);
        ɵngcc0.ɵɵelement(1, "line", 1);
        ɵngcc0.ɵɵelement(2, "line", 2);
        ɵngcc0.ɵɵelementEnd();
    } }, encapsulation: 2 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JoyrideCloseButtonComponent, [{
        type: Component,
        args: [{
                selector: 'joy-close-button',
                template: `<svg viewBox="0 0 25 25" xmlns="http://www.w3.org/2000/svg">
                <line x1="1" y1="24" 
                    x2="24" y2="1" 
                    stroke="black" 
                    stroke-width="3"/>
                <line x1="1" y1="1" 
                    x2="24" y2="24" 
                    stroke="black" 
                    stroke-width="3"/>
            </svg>`
            }]
    }], null, null); })();
