// ngx-masonry/esm2015/lib/ngx-masonry.module.js
import { NgModule } from '@angular/core';
import { NgxMasonryComponent } from './ngx-masonry.component';
import { NgxMasonryDirective } from './ngx-masonry.directive';
import * as ɵngcc0 from '@angular/core';
export class NgxMasonryModule {
}
NgxMasonryModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxMasonryModule });
NgxMasonryModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxMasonryModule_Factory(t) { return new (t || NgxMasonryModule)(); }, imports: [[]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxMasonryModule, { declarations: function () { return [NgxMasonryComponent, NgxMasonryDirective]; }, exports: function () { return [NgxMasonryComponent, NgxMasonryDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxMasonryModule, [{
        type: NgModule,
        args: [{
                imports: [],
                declarations: [NgxMasonryComponent, NgxMasonryDirective],
                exports: [NgxMasonryComponent, NgxMasonryDirective]
            }]
    }], null, null); })();
