// ngx-joyride/esm2015/lib/components/button/button.component.js
import { Component, Input, EventEmitter, Output } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = function (a0, a1, a2) { return { "background-color": a0, "color": a1, "border-color": a2 }; };
const _c1 = ["*"];
export class JoyrideButtonComponent {
    constructor() {
        this.clicked = new EventEmitter();
    }
    onClick() {
        this.clicked.emit();
    }
}
JoyrideButtonComponent.ɵfac = function JoyrideButtonComponent_Factory(t) { return new (t || JoyrideButtonComponent)(); };
JoyrideButtonComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: JoyrideButtonComponent, selectors: [["joyride-button"]], inputs: { color: "color" }, outputs: { clicked: "clicked" }, ngContentSelectors: _c1, decls: 2, vars: 5, consts: [[1, "joyride-button", 3, "ngStyle", "mouseleave", "mouseover", "click"]], template: function JoyrideButtonComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "button", 0);
        ɵngcc0.ɵɵlistener("mouseleave", function JoyrideButtonComponent_Template_button_mouseleave_0_listener() { return ctx.hover = false; })("mouseover", function JoyrideButtonComponent_Template_button_mouseover_0_listener() { return ctx.hover = true; })("click", function JoyrideButtonComponent_Template_button_click_0_listener() { return ctx.onClick(); });
        ɵngcc0.ɵɵprojection(1);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngStyle", ɵngcc0.ɵɵpureFunction3(1, _c0, ctx.hover ? "#fff" : ctx.color, ctx.hover ? ctx.color : "#fff", ctx.hover ? ctx.color : "transparent"));
    } }, directives: [ɵngcc1.NgStyle], styles: [".joyride-button[_ngcontent-%COMP%]{text-transform:uppercase;border:2px solid transparent;outline:none;padding:6px 12px;font-size:12px;font-weight:700;color:#fff;background-color:#3b5560;cursor:pointer}.joyride-button[_ngcontent-%COMP%]:hover{color:#3b5560;border:2px solid #3b5560;background-color:#fff}"] });
JoyrideButtonComponent.propDecorators = {
    color: [{ type: Input }],
    clicked: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JoyrideButtonComponent, [{
        type: Component,
        args: [{
                selector: 'joyride-button',
                template: "<button (mouseleave)=\"hover=false\" (mouseover)=\"hover=true\"\r\n    [ngStyle]=\"{'background-color': hover ? '#fff' : color, \r\n                'color': hover ? color : '#fff',\r\n                'border-color' : hover ? color : 'transparent'}\"\r\n    class=\"joyride-button\" (click)=\"onClick()\">\r\n    <ng-content></ng-content>\r\n</button>",
                styles: [".joyride-button{text-transform:uppercase;border:2px solid transparent;outline:none;padding:6px 12px;font-size:12px;font-weight:700;color:#fff;background-color:#3b5560;cursor:pointer}.joyride-button:hover{color:#3b5560;border:2px solid #3b5560;background-color:#fff}"]
            }]
    }], function () { return []; }, { clicked: [{
            type: Output
        }], color: [{
            type: Input
        }] }); })();
