// ng2-google-charts/esm2015/lib/ng2-google-charts.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { GoogleChartComponent } from './google-chart/google-chart.component';
import { GoogleChartsLoaderService } from './google-charts-loader.service';
import { GoogleChartsDashboardComponent } from './google-charts-dashboard/google-charts-dashboard.component';
import { GoogleChartsControlComponent } from './google-charts-control/google-charts-control.component';
import * as ɵngcc0 from '@angular/core';
let Ng2GoogleChartsModule = class Ng2GoogleChartsModule {
};
Ng2GoogleChartsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: Ng2GoogleChartsModule });
Ng2GoogleChartsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function Ng2GoogleChartsModule_Factory(t) { return new (t || Ng2GoogleChartsModule)(); }, providers: [
        GoogleChartsLoaderService
    ] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(Ng2GoogleChartsModule, { declarations: function () { return [GoogleChartComponent,
        GoogleChartsDashboardComponent,
        GoogleChartsControlComponent]; }, exports: function () { return [GoogleChartComponent,
        GoogleChartsDashboardComponent,
        GoogleChartsControlComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(Ng2GoogleChartsModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    GoogleChartComponent,
                    GoogleChartsDashboardComponent,
                    GoogleChartsControlComponent,
                ],
                providers: [
                    GoogleChartsLoaderService
                ],
                exports: [
                    GoogleChartComponent,
                    GoogleChartsDashboardComponent,
                    GoogleChartsControlComponent,
                ]
            }]
    }], null, null); })();
export { Ng2GoogleChartsModule };
