// ngx-editor/esm2015/lib/modules/menu/menu.component.js
import { Component, Input, ViewEncapsulation } from '@angular/core';
import { SharedService } from '../../services/shared/shared.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../services/shared/shared.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './simple-command/simple-command.component';
import * as ɵngcc4 from './link/link.component';
import * as ɵngcc5 from './image/image.component';
import * as ɵngcc6 from './dropdown/dropdown.component';
import * as ɵngcc7 from './color-picker/color-picker.component';

function MenuComponent_ng_container_1_ng_container_1_ngx_simple_command_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ngx-simple-command", 7);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r7 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(ctx_r7.iconContainerClass);
    ɵngcc0.ɵɵproperty("name", item_r5);
} }
function MenuComponent_ng_container_1_ng_container_1_ngx_link_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ngx-link", 7);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r8 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(ctx_r8.iconContainerClass);
    ɵngcc0.ɵɵproperty("name", item_r5);
} }
function MenuComponent_ng_container_1_ng_container_1_ngx_image_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ngx-image", 7);
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r9 = ɵngcc0.ɵɵnextContext(2);
    ɵngcc0.ɵɵclassMap(ctx_r9.iconContainerClass);
    ɵngcc0.ɵɵproperty("name", item_r5);
} }
function MenuComponent_ng_container_1_ng_container_1_ng_container_4_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelement(1, "ngx-dropdown", 8);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const dropdown_r18 = ctx.$implicit;
    const ctx_r17 = ɵngcc0.ɵɵnextContext(4);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵclassMap(ctx_r17.dropdownContainerClass);
    ɵngcc0.ɵɵproperty("group", dropdown_r18.key)("items", dropdown_r18.value);
} }
function MenuComponent_ng_container_1_ng_container_1_ng_container_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, MenuComponent_ng_container_1_ng_container_1_ng_container_4_ng_container_1_Template, 2, 4, "ng-container", 1);
    ɵngcc0.ɵɵpipe(2, "keyvalue");
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r5 = ɵngcc0.ɵɵnextContext().$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", ɵngcc0.ɵɵpipeBind1(2, 1, item_r5));
} }
function MenuComponent_ng_container_1_ng_container_1_ngx_color_picker_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ngx-color-picker", 9);
} if (rf & 2) {
    const ctx_r11 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassMap(ctx_r11.iconContainerClass);
} }
function MenuComponent_ng_container_1_ng_container_1_ngx_color_picker_6_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "ngx-color-picker", 10);
} if (rf & 2) {
    const ctx_r12 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassMap(ctx_r12.iconContainerClass);
} }
function MenuComponent_ng_container_1_ng_container_1_div_7_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div");
} if (rf & 2) {
    const ctx_r13 = ɵngcc0.ɵɵnextContext(3);
    ɵngcc0.ɵɵclassMap(ctx_r13.seperatorClass);
} }
function MenuComponent_ng_container_1_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, MenuComponent_ng_container_1_ng_container_1_ngx_simple_command_1_Template, 1, 3, "ngx-simple-command", 3);
    ɵngcc0.ɵɵtemplate(2, MenuComponent_ng_container_1_ng_container_1_ngx_link_2_Template, 1, 3, "ngx-link", 3);
    ɵngcc0.ɵɵtemplate(3, MenuComponent_ng_container_1_ng_container_1_ngx_image_3_Template, 1, 3, "ngx-image", 3);
    ɵngcc0.ɵɵtemplate(4, MenuComponent_ng_container_1_ng_container_1_ng_container_4_Template, 3, 3, "ng-container", 2);
    ɵngcc0.ɵɵtemplate(5, MenuComponent_ng_container_1_ng_container_1_ngx_color_picker_5_Template, 1, 2, "ngx-color-picker", 4);
    ɵngcc0.ɵɵtemplate(6, MenuComponent_ng_container_1_ng_container_1_ngx_color_picker_6_Template, 1, 2, "ngx-color-picker", 5);
    ɵngcc0.ɵɵtemplate(7, MenuComponent_ng_container_1_ng_container_1_div_7_Template, 1, 2, "div", 6);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const item_r5 = ctx.$implicit;
    const lastItem_r6 = ctx.last;
    const lastToolbarItem_r3 = ɵngcc0.ɵɵnextContext().last;
    const ctx_r4 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.simpleCommands.includes(item_r5));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5 === "link");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5 === "image");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r4.isDropDown(item_r5));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5 === "text_color");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", item_r5 === "background_color");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", lastItem_r6 && !lastToolbarItem_r3);
} }
function MenuComponent_ng_container_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵtemplate(1, MenuComponent_ng_container_1_ng_container_1_Template, 8, 7, "ng-container", 1);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const toolbarItem_r2 = ctx.$implicit;
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngForOf", toolbarItem_r2);
} }
function MenuComponent_ng_container_2_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementContainer(1, 11);
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", ctx_r1.customMenuRef);
} }
const _c0 = function (a0) { return { "NgxEditor--Disabled": a0 }; };
export class MenuComponent {
    constructor(sharedService) {
        this.sharedService = sharedService;
        this.disabled = false;
        this.customMenuRef = null;
        this.simpleCommands = [
            'bold', 'italic',
            'underline', 'strike',
            'code', 'blockquote',
            'ordered_list', 'bullet_list',
            'align_left', 'align_center', 'align_right', 'align_justify'
        ];
        this.iconContainerClass = ['NgxEditor__MenuItem', 'NgxEditor__MenuItem--Icon'];
        this.dropdownContainerClass = ['NgxEditor__Dropdown'];
        this.seperatorClass = ['NgxEditor__Seperator'];
        this.customMenuRefSubscription = this.sharedService.customMenuRefChange.subscribe((ref) => {
            this.customMenuRef = ref;
        });
    }
    isDropDown(item) {
        if (typeof item === 'object') {
            return true;
        }
        return false;
    }
    ngOnDestroy() {
        this.customMenuRefSubscription.unsubscribe();
    }
}
MenuComponent.ɵfac = function MenuComponent_Factory(t) { return new (t || MenuComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.SharedService)); };
MenuComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: MenuComponent, selectors: [["ngx-menu"]], inputs: { disabled: "disabled", toolbar: "toolbar", editorView: "editorView" }, decls: 3, vars: 5, consts: [[1, "NgxEditor__MenuBar", 3, "ngClass"], [4, "ngFor", "ngForOf"], [4, "ngIf"], [3, "name", "class", 4, "ngIf"], ["type", "text_color", 3, "class", 4, "ngIf"], ["type", "background_color", 3, "class", 4, "ngIf"], [3, "class", 4, "ngIf"], [3, "name"], [3, "group", "items"], ["type", "text_color"], ["type", "background_color"], [3, "ngTemplateOutlet"]], template: function MenuComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵtemplate(1, MenuComponent_ng_container_1_Template, 2, 1, "ng-container", 1);
        ɵngcc0.ɵɵtemplate(2, MenuComponent_ng_container_2_Template, 2, 1, "ng-container", 2);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction1(3, _c0, ctx.disabled));
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.toolbar);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.customMenuRef);
    } }, directives: [ɵngcc2.NgClass, ɵngcc2.NgForOf, ɵngcc2.NgIf, ɵngcc3.SimpleCommandComponent, ɵngcc4.LinkComponent, ɵngcc5.ImageComponent, ɵngcc6.DropdownComponent, ɵngcc7.ColorPickerComponent, ɵngcc2.NgTemplateOutlet], pipes: [ɵngcc2.KeyValuePipe], styles: [".NgxEditor__MenuBar{border-bottom:1px solid #ddd;cursor:default;display:flex;height:1.85rem;padding:.2rem}.NgxEditor__MenuItem{align-items:center;border-radius:2px;display:flex;flex-shrink:0;justify-content:center;position:relative}.NgxEditor__MenuItem:hover{background-color:#f1f1f1}.NgxEditor__MenuItem.NgxEditor__MenuItem--Icon{height:1.85rem;transition:.3s ease-in-out;width:1.85rem}.NgxEditor__MenuItem.NgxEditor__MenuItem--Icon+.NgxEditor__MenuItem--Icon{margin-left:2px}.NgxEditor__MenuItem .NgxEditor__MenuItem--IconContainer{align-items:center;display:flex;height:100%;justify-content:center;width:100%}.NgxEditor__MenuItem.NgxEditor__MenuItem--Text{padding:0 .3rem}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active{background-color:#e8f0fe}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active.NgxEditor__MenuItem--Text{color:#1a73e8}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active svg{fill:#1a73e8}.NgxEditor__Dropdown{align-items:center;display:flex;flex-shrink:0;min-width:4rem;position:relative}.NgxEditor__Dropdown:hover{background-color:#f1f1f1}.NgxEditor__Dropdown .NgxEditor__Dropdown--Text{align-items:center;display:flex;height:100%;justify-content:center;padding:0 .3rem;width:100%}.NgxEditor__Dropdown .NgxEditor__Dropdown--Text:after{border-bottom:0;border-left:.25rem solid transparent;border-right:.25rem solid transparent;border-top:.25rem solid;content:\"\";display:inline-block;margin-left:1.5rem;vertical-align:.25rem}.NgxEditor__Dropdown .NgxEditor__Dropdown--DropdownMenu{background-color:#fff;border-radius:4px;box-shadow:0 2px 6px 2px rgba(60,64,67,.15);left:0;position:absolute;top:calc(1.85rem + 2px);width:100%;z-index:10}.NgxEditor__Dropdown .NgxEditor__Dropdown--Item{color:inherit;padding:.5rem;white-space:nowrap}.NgxEditor__Dropdown .NgxEditor__Dropdown--Item:hover{background-color:#ececec}.NgxEditor__Dropdown.NgxEditor__Dropdown--Open,.NgxEditor__Dropdown.NgxEditor__Dropdown--Selected{background-color:#e8f0fe}.NgxEditor__Dropdown.NgxEditor__Dropdown--Open .NgxEditor__Dropdown--Text,.NgxEditor__Dropdown.NgxEditor__Dropdown--Selected .NgxEditor__Dropdown--Text{color:#1a73e8}.NgxEditor__Dropdown .NgxEditor__Dropdown--Active{background-color:#f1f1f1}.NgxEditor__Dropdown .NgxEditor__Dropdown--Active:hover{background-color:#e7e7e7}.NgxEditor__Popup{background-color:#fff;border-radius:4px;box-shadow:0 2px 6px 2px rgba(60,64,67,.15);min-width:12rem;padding:8px;position:absolute;top:calc(1.85rem + 2px);z-index:10}.NgxEditor__Popup .NgxEditor__Popup--FormGroup{margin-bottom:8px}.NgxEditor__Popup .NgxEditor__Popup--FormGroup label{margin-bottom:3px}.NgxEditor__Popup .NgxEditor__Popup--FormGroup input[type=text],.NgxEditor__Popup .NgxEditor__Popup--FormGroup input[type=url]{padding:2px 4px}.NgxEditor__Popup .NgxEditor__Popup--Col{display:flex;flex-direction:column;position:relative}.NgxEditor__Popup .NgxEditor__Popup--Label{font-size:85%}.NgxEditor__Seperator{border-left:1px solid #ccc;margin:0 .3rem}"], encapsulation: 2 });
MenuComponent.ctorParameters = () => [
    { type: SharedService }
];
MenuComponent.propDecorators = {
    toolbar: [{ type: Input }],
    editorView: [{ type: Input }],
    disabled: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MenuComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-menu',
                template: "<div class=\"NgxEditor__MenuBar\" [ngClass]=\"{'NgxEditor--Disabled': disabled}\">\n\n  <ng-container *ngFor=\"let toolbarItem of toolbar; let lastToolbarItem = last\">\n    <ng-container *ngFor=\"let item of toolbarItem; let lastItem = last\">\n\n      <!-- simple icons -->\n      <ngx-simple-command [name]=\"item\" [class]=\"iconContainerClass\" *ngIf=\"simpleCommands.includes(item)\">\n      </ngx-simple-command>\n\n      <!-- link -->\n      <ngx-link [name]=\"item\" [class]=\"iconContainerClass\" *ngIf=\"item === 'link'\">\n      </ngx-link>\n\n      <!-- image -->\n      <ngx-image [name]=\"item\" [class]=\"iconContainerClass\" *ngIf=\"item === 'image'\">\n      </ngx-image>\n\n      <!-- dropdown -->\n      <ng-container *ngIf=\"isDropDown(item)\">\n        <ng-container *ngFor=\"let dropdown of item | keyvalue\">\n          <ngx-dropdown [class]=\"dropdownContainerClass\" [group]=\"dropdown.key\" [items]=\"dropdown.value\">\n          </ngx-dropdown>\n        </ng-container>\n      </ng-container>\n\n      <!-- text color picker -->\n      <ngx-color-picker [class]=\"iconContainerClass\" *ngIf=\"item === 'text_color'\" type=\"text_color\">\n      </ngx-color-picker>\n      <!-- background color picker -->\n      <ngx-color-picker [class]=\"iconContainerClass\" *ngIf=\"item === 'background_color'\" type=\"background_color\">\n      </ngx-color-picker>\n\n      <!-- seperator -->\n      <div [class]=\"seperatorClass\" *ngIf=\"lastItem && !lastToolbarItem\"></div>\n    </ng-container>\n  </ng-container>\n\n  <!-- custom menu -->\n  <ng-container *ngIf=\"customMenuRef\">\n    <ng-container [ngTemplateOutlet]=\"customMenuRef\"></ng-container>\n  </ng-container>\n\n</div>\n",
                encapsulation: ViewEncapsulation.None,
                styles: [".NgxEditor__MenuBar{border-bottom:1px solid #ddd;cursor:default;display:flex;height:1.85rem;padding:.2rem}.NgxEditor__MenuItem{align-items:center;border-radius:2px;display:flex;flex-shrink:0;justify-content:center;position:relative}.NgxEditor__MenuItem:hover{background-color:#f1f1f1}.NgxEditor__MenuItem.NgxEditor__MenuItem--Icon{height:1.85rem;transition:.3s ease-in-out;width:1.85rem}.NgxEditor__MenuItem.NgxEditor__MenuItem--Icon+.NgxEditor__MenuItem--Icon{margin-left:2px}.NgxEditor__MenuItem .NgxEditor__MenuItem--IconContainer{align-items:center;display:flex;height:100%;justify-content:center;width:100%}.NgxEditor__MenuItem.NgxEditor__MenuItem--Text{padding:0 .3rem}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active{background-color:#e8f0fe}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active.NgxEditor__MenuItem--Text{color:#1a73e8}.NgxEditor__MenuItem.NgxEditor__MenuItem--Active svg{fill:#1a73e8}.NgxEditor__Dropdown{align-items:center;display:flex;flex-shrink:0;min-width:4rem;position:relative}.NgxEditor__Dropdown:hover{background-color:#f1f1f1}.NgxEditor__Dropdown .NgxEditor__Dropdown--Text{align-items:center;display:flex;height:100%;justify-content:center;padding:0 .3rem;width:100%}.NgxEditor__Dropdown .NgxEditor__Dropdown--Text:after{border-bottom:0;border-left:.25rem solid transparent;border-right:.25rem solid transparent;border-top:.25rem solid;content:\"\";display:inline-block;margin-left:1.5rem;vertical-align:.25rem}.NgxEditor__Dropdown .NgxEditor__Dropdown--DropdownMenu{background-color:#fff;border-radius:4px;box-shadow:0 2px 6px 2px rgba(60,64,67,.15);left:0;position:absolute;top:calc(1.85rem + 2px);width:100%;z-index:10}.NgxEditor__Dropdown .NgxEditor__Dropdown--Item{color:inherit;padding:.5rem;white-space:nowrap}.NgxEditor__Dropdown .NgxEditor__Dropdown--Item:hover{background-color:#ececec}.NgxEditor__Dropdown.NgxEditor__Dropdown--Open,.NgxEditor__Dropdown.NgxEditor__Dropdown--Selected{background-color:#e8f0fe}.NgxEditor__Dropdown.NgxEditor__Dropdown--Open .NgxEditor__Dropdown--Text,.NgxEditor__Dropdown.NgxEditor__Dropdown--Selected .NgxEditor__Dropdown--Text{color:#1a73e8}.NgxEditor__Dropdown .NgxEditor__Dropdown--Active{background-color:#f1f1f1}.NgxEditor__Dropdown .NgxEditor__Dropdown--Active:hover{background-color:#e7e7e7}.NgxEditor__Popup{background-color:#fff;border-radius:4px;box-shadow:0 2px 6px 2px rgba(60,64,67,.15);min-width:12rem;padding:8px;position:absolute;top:calc(1.85rem + 2px);z-index:10}.NgxEditor__Popup .NgxEditor__Popup--FormGroup{margin-bottom:8px}.NgxEditor__Popup .NgxEditor__Popup--FormGroup label{margin-bottom:3px}.NgxEditor__Popup .NgxEditor__Popup--FormGroup input[type=text],.NgxEditor__Popup .NgxEditor__Popup--FormGroup input[type=url]{padding:2px 4px}.NgxEditor__Popup .NgxEditor__Popup--Col{display:flex;flex-direction:column;position:relative}.NgxEditor__Popup .NgxEditor__Popup--Label{font-size:85%}.NgxEditor__Seperator{border-left:1px solid #ccc;margin:0 .3rem}"]
            }]
    }], function () { return [{ type: ɵngcc1.SharedService }]; }, { disabled: [{
            type: Input
        }], toolbar: [{
            type: Input
        }], editorView: [{
            type: Input
        }] }); })();
