// ngx-editor/esm2015/lib/modules/menu/simple-command/simple-command.component.js
import { Component, HostBinding, Input } from '@angular/core';
import { SimpleCommands } from '../MenuCommands';
import Icon from '../../../icons';
import { NgxEditorService } from '../../../editor.service';
import { SharedService } from '../../../services/shared/shared.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../editor.service';
import * as ɵngcc2 from '../../../services/shared/shared.service';
import * as ɵngcc3 from '../../../pipes/sanitize/sanitize-html.pipe';
export class SimpleCommandComponent {
    constructor(ngxeService, sharedService) {
        this.ngxeService = ngxeService;
        this.sharedService = sharedService;
        this.isActive = false;
        this.disabled = false;
        this.update = (view) => {
            const { state } = view;
            const command = SimpleCommands.get(this.name);
            this.isActive = command.isActive(state);
            this.disabled = !command.canExecute(state);
        };
        this.editorView = this.sharedService.view;
        this.pluginUpdateSubscription = this.sharedService.plugin.update.subscribe((view) => {
            this.update(view);
        });
    }
    toggle(e) {
        e.preventDefault();
        if (e.button !== 0) {
            return;
        }
        const { state, dispatch } = this.editorView;
        const command = SimpleCommands.get(this.name);
        command.execute(state, dispatch);
    }
    getTitle(name) {
        return this.ngxeService.locals.get(name);
    }
    ngOnInit() {
        this.html = Icon.get(this.name);
    }
    ngOnDestroy() {
        this.pluginUpdateSubscription.unsubscribe();
    }
}
SimpleCommandComponent.ɵfac = function SimpleCommandComponent_Factory(t) { return new (t || SimpleCommandComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.NgxEditorService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.SharedService)); };
SimpleCommandComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: SimpleCommandComponent, selectors: [["ngx-simple-command"]], hostVars: 4, hostBindings: function SimpleCommandComponent_HostBindings(rf, ctx) { if (rf & 2) {
        ɵngcc0.ɵɵclassProp("NgxEditor__MenuItem--Active", ctx.isActive)("NgxEditor--Disabled", ctx.disabled);
    } }, inputs: { name: "name" }, decls: 2, vars: 4, consts: [[1, "NgxEditor__MenuItem--IconContainer", 3, "innerHTML", "title", "mousedown"]], template: function SimpleCommandComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵlistener("mousedown", function SimpleCommandComponent_Template_div_mousedown_0_listener($event) { return ctx.toggle($event); });
        ɵngcc0.ɵɵpipe(1, "sanitizeHtml");
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("innerHTML", ɵngcc0.ɵɵpipeBind1(1, 2, ctx.html), ɵngcc0.ɵɵsanitizeHtml)("title", ctx.getTitle(ctx.name));
    } }, pipes: [ɵngcc3.SanitizeHtmlPipe], styles: [""] });
SimpleCommandComponent.ctorParameters = () => [
    { type: NgxEditorService },
    { type: SharedService }
];
SimpleCommandComponent.propDecorators = {
    name: [{ type: Input }],
    isActive: [{ type: HostBinding, args: ['class.NgxEditor__MenuItem--Active',] }],
    disabled: [{ type: HostBinding, args: ['class.NgxEditor--Disabled',] }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SimpleCommandComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-simple-command',
                template: "<div class=\"NgxEditor__MenuItem--IconContainer\" [innerHTML]=\"html | sanitizeHtml\" (mousedown)=\"toggle($event)\"\n  [title]=\"getTitle(name)\">\n</div>\n",
                styles: [""]
            }]
    }], function () { return [{ type: ɵngcc1.NgxEditorService }, { type: ɵngcc2.SharedService }]; }, { isActive: [{
            type: HostBinding,
            args: ['class.NgxEditor__MenuItem--Active']
        }], disabled: [{
            type: HostBinding,
            args: ['class.NgxEditor--Disabled']
        }], name: [{
            type: Input
        }] }); })();
