// @ks89/angular-modal-gallery/esm5/lib/directives/a-tag-bg-image.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
import { Image } from '../model/image.class';
/**
 * Directive to add an image to an `<a>` tag with some additional custom properties.
 */
import * as ɵngcc0 from '@angular/core';
var ATagBgImageDirective = /** @class */ (function () {
    function ATagBgImageDirective(renderer, el) {
        this.renderer = renderer;
        this.el = el;
    }
    /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     */
    /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    ATagBgImageDirective.prototype.ngOnInit = /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    function () {
        this.applyStyle();
    };
    /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     */
    /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     * @return {?}
     */
    ATagBgImageDirective.prototype.ngOnChanges = /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     * @return {?}
     */
    function () {
        this.applyStyle();
    };
    /**
     * Private method to add an image as background of an `<a>` tag.
     */
    /**
     * Private method to add an image as background of an `<a>` tag.
     * @private
     * @return {?}
     */
    ATagBgImageDirective.prototype.applyStyle = /**
     * Private method to add an image as background of an `<a>` tag.
     * @private
     * @return {?}
     */
    function () {
        if (!this.image || (!this.image.plain && !this.image.modal)) {
            return;
        }
        /** @type {?} */
        var imgPath = this.image.plain && this.image.plain.img ? this.image.plain.img : this.image.modal.img;
        this.renderer.setStyle(this.el.nativeElement, 'background', "url(\"" + imgPath + "\") " + this.style);
    };
    /** @nocollapse */
    ATagBgImageDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    ATagBgImageDirective.propDecorators = {
        image: [{ type: Input }],
        style: [{ type: Input }]
    };
ATagBgImageDirective.ɵfac = function ATagBgImageDirective_Factory(t) { return new (t || ATagBgImageDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
ATagBgImageDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: ATagBgImageDirective, selectors: [["", "ksATagBgImage", ""]], inputs: { image: "image", style: "style" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ATagBgImageDirective, [{
        type: Directive,
        args: [{
                selector: '[ksATagBgImage]'
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { image: [{
            type: Input
        }], style: [{
            type: Input
        }] }); })();
    return ATagBgImageDirective;
}());
export { ATagBgImageDirective };
if (false) {
    /**
     * Object of type `Image` that represents the image to add to the `<a>` tag.
     * @type {?}
     */
    ATagBgImageDirective.prototype.image;
    /**
     * Additional style to customize the background attribute.
     * Empty string by default.
     * @type {?}
     */
    ATagBgImageDirective.prototype.style;
    /**
     * @type {?}
     * @private
     */
    ATagBgImageDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    ATagBgImageDirective.prototype.el;
}
