// @ks89/angular-modal-gallery/esm5/lib/directives/keyboard-navigation.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { Directive, EventEmitter, HostListener, Input, Output } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
var KeyboardNavigationDirective = /** @class */ (function () {
    function KeyboardNavigationDirective() {
        this.keyPress = new EventEmitter();
    }
    /**
     * Listener to catch keyboard's events and call the right method based on the key.
     * For instance, pressing esc, this will call `closeGallery(Action.KEYBOARD)` and so on.
     * If you passed a valid `keyboardConfig` esc, right and left buttons will be customized based on your data.
     * @param e KeyboardEvent caught by the listener.
     */
    /**
     * Listener to catch keyboard's events and call the right method based on the key.
     * For instance, pressing esc, this will call `closeGallery(Action.KEYBOARD)` and so on.
     * If you passed a valid `keyboardConfig` esc, right and left buttons will be customized based on your data.
     * @param {?} e KeyboardEvent caught by the listener.
     * @return {?}
     */
    KeyboardNavigationDirective.prototype.onKeyDown = /**
     * Listener to catch keyboard's events and call the right method based on the key.
     * For instance, pressing esc, this will call `closeGallery(Action.KEYBOARD)` and so on.
     * If you passed a valid `keyboardConfig` esc, right and left buttons will be customized based on your data.
     * @param {?} e KeyboardEvent caught by the listener.
     * @return {?}
     */
    function (e) {
        if (!this.isOpen) {
            return;
        }
        this.keyPress.emit(e.keyCode);
    };
    KeyboardNavigationDirective.propDecorators = {
        isOpen: [{ type: Input }],
        keyPress: [{ type: Output }],
        onKeyDown: [{ type: HostListener, args: ['window:keydown', ['$event'],] }]
    };
KeyboardNavigationDirective.ɵfac = function KeyboardNavigationDirective_Factory(t) { return new (t || KeyboardNavigationDirective)(); };
KeyboardNavigationDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: KeyboardNavigationDirective, selectors: [["", "ksKeyboardNavigation", ""]], hostBindings: function KeyboardNavigationDirective_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function KeyboardNavigationDirective_keydown_HostBindingHandler($event) { return ctx.onKeyDown($event); }, false, ɵngcc0.ɵɵresolveWindow);
    } }, inputs: { isOpen: "isOpen" }, outputs: { keyPress: "keyPress" } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(KeyboardNavigationDirective, [{
        type: Directive,
        args: [{
                selector: '[ksKeyboardNavigation]'
            }]
    }], function () { return []; }, { keyPress: [{
            type: Output
        }], onKeyDown: [{
            type: HostListener,
            args: ['window:keydown', ['$event']]
        }], isOpen: [{
            type: Input
        }] }); })();
    return KeyboardNavigationDirective;
}());
export { KeyboardNavigationDirective };
if (false) {
    /** @type {?} */
    KeyboardNavigationDirective.prototype.isOpen;
    /** @type {?} */
    KeyboardNavigationDirective.prototype.keyPress;
}
