// @ks89/angular-modal-gallery/esm5/lib/directives/max-size.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { Directive, ElementRef, Input, Renderer2 } from '@angular/core';
/**
 * Directive to change the max size of an element.
 */
import * as ɵngcc0 from '@angular/core';
var MaxSizeDirective = /** @class */ (function () {
    function MaxSizeDirective(renderer, el) {
        this.renderer = renderer;
        this.el = el;
    }
    /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     */
    /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    MaxSizeDirective.prototype.ngOnInit = /**
     * Method ´ngOnInit´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    function () {
        this.applyStyle();
    };
    /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     */
    /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     * @return {?}
     */
    MaxSizeDirective.prototype.ngOnChanges = /**
     * Method ´ngOnChanges´ to apply the style of this directive.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called when any data-bound property of a directive changes!!!
     * @return {?}
     */
    function () {
        this.applyStyle();
    };
    /**
     * Private method to change both max-width and max-height of an element.
     */
    /**
     * Private method to change both max-width and max-height of an element.
     * @private
     * @return {?}
     */
    MaxSizeDirective.prototype.applyStyle = /**
     * Private method to change both max-width and max-height of an element.
     * @private
     * @return {?}
     */
    function () {
        if (!this.sizeConfig) {
            return;
        }
        if (this.sizeConfig.maxWidth) {
            this.renderer.setStyle(this.el.nativeElement, 'max-width', this.sizeConfig.maxWidth);
        }
        if (this.sizeConfig.maxHeight) {
            this.renderer.setStyle(this.el.nativeElement, 'max-height', this.sizeConfig.maxHeight);
        }
    };
    /** @nocollapse */
    MaxSizeDirective.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    MaxSizeDirective.propDecorators = {
        sizeConfig: [{ type: Input }]
    };
MaxSizeDirective.ɵfac = function MaxSizeDirective_Factory(t) { return new (t || MaxSizeDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
MaxSizeDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: MaxSizeDirective, selectors: [["", "ksMaxSize", ""]], inputs: { sizeConfig: "sizeConfig" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MaxSizeDirective, [{
        type: Directive,
        args: [{
                selector: '[ksMaxSize]'
            }]
    }], function () { return [{ type: ɵngcc0.Renderer2 }, { type: ɵngcc0.ElementRef }]; }, { sizeConfig: [{
            type: Input
        }] }); })();
    return MaxSizeDirective;
}());
export { MaxSizeDirective };
if (false) {
    /**
     * Object of type `MaxSize` to resize the element.
     * @type {?}
     */
    MaxSizeDirective.prototype.sizeConfig;
    /**
     * @type {?}
     * @private
     */
    MaxSizeDirective.prototype.renderer;
    /**
     * @type {?}
     * @private
     */
    MaxSizeDirective.prototype.el;
}
