// ngx-joyride/esm2015/lib/components/arrow/arrow.component.js
import { Component, Input, ViewEncapsulation } from "@angular/core";
import * as ɵngcc0 from '@angular/core';
export class JoyrideArrowComponent {
    constructor() {
        this.position = 'top';
    }
}
JoyrideArrowComponent.ɵfac = function JoyrideArrowComponent_Factory(t) { return new (t || JoyrideArrowComponent)(); };
JoyrideArrowComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: JoyrideArrowComponent, selectors: [["joyride-arrow"]], inputs: { position: "position" }, decls: 1, vars: 8, template: function JoyrideArrowComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelement(0, "div");
    } if (rf & 2) {
        ɵngcc0.ɵɵclassProp("joyride-arrow__top", ctx.position == "top")("joyride-arrow__bottom", ctx.position == "bottom")("joyride-arrow__left", ctx.position == "left")("joyride-arrow__right", ctx.position == "right");
    } }, styles: [".joyride-arrow__top{border-bottom:11px solid #fff}.joyride-arrow__bottom,.joyride-arrow__top{border-left:11px solid transparent;border-right:11px solid transparent}.joyride-arrow__bottom{border-top:11px solid #fff}.joyride-arrow__right{border-left:11px solid #fff}.joyride-arrow__left,.joyride-arrow__right{border-bottom:11px solid transparent;border-top:11px solid transparent}.joyride-arrow__left{border-right:11px solid #fff}"], encapsulation: 2 });
JoyrideArrowComponent.propDecorators = {
    position: [{ type: Input }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JoyrideArrowComponent, [{
        type: Component,
        args: [{
                selector: 'joyride-arrow',
                template: "<div [class.joyride-arrow__top]=\"position == 'top'\"\r\n     [class.joyride-arrow__bottom]=\"position == 'bottom'\"\r\n     [class.joyride-arrow__left]=\"position == 'left'\"\r\n     [class.joyride-arrow__right]=\"position == 'right'\">\r\n</div>",
                encapsulation: ViewEncapsulation.None,
                styles: [".joyride-arrow__top{border-bottom:11px solid #fff}.joyride-arrow__bottom,.joyride-arrow__top{border-left:11px solid transparent;border-right:11px solid transparent}.joyride-arrow__bottom{border-top:11px solid #fff}.joyride-arrow__right{border-left:11px solid #fff}.joyride-arrow__left,.joyride-arrow__right{border-bottom:11px solid transparent;border-top:11px solid transparent}.joyride-arrow__left{border-right:11px solid #fff}"]
            }]
    }], function () { return []; }, { position: [{
            type: Input
        }] }); })();
