// ngx-editor/esm2015/lib/editor.module.js
import { NgModule, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxEditorComponent } from './editor.component';
import { NgxEditorServiceConfig, provideMyServiceOptions } from './editor.service';
import { MenuModule } from './modules/menu/menu.module';
import { BubbleComponent } from './components/bubble/bubble.component';
import { SharedService } from './services/shared/shared.service';
import * as ɵngcc0 from '@angular/core';
const NGX_EDITOR_CONFIG_TOKEN = new InjectionToken('NgxEditorConfig');
export class NgxEditorModule {
    static forRoot(config) {
        return {
            ngModule: NgxEditorModule,
            providers: [
                {
                    provide: NGX_EDITOR_CONFIG_TOKEN,
                    useValue: config
                },
                {
                    provide: NgxEditorServiceConfig,
                    useFactory: provideMyServiceOptions,
                    deps: [NGX_EDITOR_CONFIG_TOKEN]
                }
            ]
        };
    }
}
NgxEditorModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgxEditorModule });
NgxEditorModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function NgxEditorModule_Factory(t) { return new (t || NgxEditorModule)(); }, providers: [
        SharedService
    ], imports: [[
            CommonModule,
            MenuModule,
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxEditorModule, { declarations: function () { return [NgxEditorComponent,
        BubbleComponent]; }, imports: function () { return [CommonModule,
        MenuModule]; }, exports: function () { return [NgxEditorComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(NgxEditorModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MenuModule,
                ],
                providers: [
                    SharedService
                ],
                declarations: [
                    NgxEditorComponent,
                    BubbleComponent
                ],
                exports: [NgxEditorComponent],
                entryComponents: [BubbleComponent]
            }]
    }], null, null); })();
