// ngx-editor/esm2015/lib/components/bubble/bubble.component.js
import { Component, ElementRef, Renderer2 } from '@angular/core';
import { NodeSelection } from 'prosemirror-state';
import { calculateBubblePos, getSelectionMarks, isLinkActive } from 'ngx-editor/helpers';
import { removeLink } from 'ngx-editor/commands';
import { SharedService } from '../../services/shared/shared.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../services/shared/shared.service';
import * as ɵngcc2 from '@angular/common';

function BubbleComponent_ng_container_0_Template(rf, ctx) { if (rf & 1) {
    const _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "a", 1);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(3, "div", 2);
    ɵngcc0.ɵɵelementStart(4, "button", 3);
    ɵngcc0.ɵɵlistener("click", function BubbleComponent_ng_container_0_Template_button_click_4_listener() { ɵngcc0.ɵɵrestoreView(_r2); const ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.removeLink(); });
    ɵngcc0.ɵɵtext(5, "Remove link");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("href", ctx_r0.activeLinkItem.attrs.href, ɵngcc0.ɵɵsanitizeUrl)("title", ctx_r0.activeLinkItem.attrs.href);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r0.activeLinkItem.attrs.href, " ");
} }
export class BubbleComponent {
    constructor(sharedService, el, renderer) {
        this.sharedService = sharedService;
        this.el = el;
        this.renderer = renderer;
        this.pluginUpdateSubscription = this.sharedService.plugin.update.subscribe((view) => {
            this.view = view;
            this.update(view);
        });
    }
    setDomPosition(view) {
        // Otherwise, reposition it and update its content
        this.showBubble();
        const { bottom, left } = calculateBubblePos(view, this.el.nativeElement);
        this.renderer.setStyle(this.el.nativeElement, 'left', `${left}px`);
        this.renderer.setStyle(this.el.nativeElement, 'bottom', `${bottom}px`);
    }
    showBubble() {
        this.renderer.setStyle(this.el.nativeElement, 'display', '');
    }
    hideBubble() {
        this.renderer.setStyle(this.el.nativeElement, 'display', 'none');
    }
    removeLink() {
        const { state, dispatch } = this.view;
        removeLink()(state, dispatch);
        this.view.focus();
    }
    update(view) {
        const { state } = view;
        const { schema, selection } = state;
        if (!schema.marks.link) {
            return;
        }
        if (selection instanceof NodeSelection) {
            if (selection.node.type.name === 'image') {
                return;
            }
        }
        const hasFocus = view.hasFocus();
        const isActive = isLinkActive(state);
        const linkMarks = getSelectionMarks(state).filter(mark => mark.type === schema.marks.link);
        // hide for selection and show only for clicks
        if (!hasFocus || !isActive) {
            this.hideBubble();
            return;
        }
        const [linkItem] = linkMarks;
        this.activeLinkItem = linkItem;
        // update dom position
        this.setDomPosition(view);
    }
    ngOnInit() {
    }
    ngOnDestroy() {
        this.pluginUpdateSubscription.unsubscribe();
    }
}
BubbleComponent.ɵfac = function BubbleComponent_Factory(t) { return new (t || BubbleComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.SharedService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.Renderer2)); };
BubbleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BubbleComponent, selectors: [["ngx-bubble"]], decls: 1, vars: 1, consts: [[4, "ngIf"], ["target", "_blank", 3, "href", "title"], [1, "commands"], [1, "command", 3, "click"]], template: function BubbleComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, BubbleComponent_ng_container_0_Template, 6, 3, "ng-container", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.activeLinkItem);
    } }, directives: [ɵngcc2.NgIf], styles: ["[_nghost-%COMP%]{align-items:center;background:#fff;border:1px solid silver;border-radius:4px;display:flex;margin-bottom:.3rem;padding:.3rem;position:absolute;transform:translateX(-50%);z-index:20}[_nghost-%COMP%]:after, [_nghost-%COMP%]:before{border:solid transparent;border-width:5px 5px 0;content:\"\";height:0;left:50%;margin-left:-5px;position:absolute;width:0}[_nghost-%COMP%]:before{border-top-color:silver;bottom:-6px}[_nghost-%COMP%]:after{border-top-color:#fff;bottom:-4.5px}[_nghost-%COMP%]   a[_ngcontent-%COMP%]{display:inline-block;margin-left:2px;max-width:15rem;overflow:hidden;text-overflow:ellipsis}[_nghost-%COMP%]   .commands[_ngcontent-%COMP%]{display:flex;margin-left:.5rem}[_nghost-%COMP%]   .commands[_ngcontent-%COMP%]   .command[_ngcontent-%COMP%]{background-color:#f1f1f1;border:none;border-radius:2px;display:flex;padding:2px;white-space:nowrap}[_nghost-%COMP%]   .commands[_ngcontent-%COMP%]   .command[_ngcontent-%COMP%]:hover{text-decoration:underline}"] });
BubbleComponent.ctorParameters = () => [
    { type: SharedService },
    { type: ElementRef },
    { type: Renderer2 }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BubbleComponent, [{
        type: Component,
        args: [{
                selector: 'ngx-bubble',
                template: "<ng-container *ngIf=\"activeLinkItem\">\n  <a [href]=\"activeLinkItem.attrs.href\" target=\"_blank\" [title]=\"activeLinkItem.attrs.href\">\n    {{activeLinkItem.attrs.href}}\n  </a>\n  <div class=\"commands\">\n    <button class=\"command\" (click)=\"removeLink()\">Remove link</button>\n  </div>\n</ng-container>\n",
                styles: [":host{align-items:center;background:#fff;border:1px solid silver;border-radius:4px;display:flex;margin-bottom:.3rem;padding:.3rem;position:absolute;transform:translateX(-50%);z-index:20}:host:after,:host:before{border:solid transparent;border-width:5px 5px 0;content:\"\";height:0;left:50%;margin-left:-5px;position:absolute;width:0}:host:before{border-top-color:silver;bottom:-6px}:host:after{border-top-color:#fff;bottom:-4.5px}:host a{display:inline-block;margin-left:2px;max-width:15rem;overflow:hidden;text-overflow:ellipsis}:host .commands{display:flex;margin-left:.5rem}:host .commands .command{background-color:#f1f1f1;border:none;border-radius:2px;display:flex;padding:2px;white-space:nowrap}:host .commands .command:hover{text-decoration:underline}"]
            }]
    }], function () { return [{ type: ɵngcc1.SharedService }, { type: ɵngcc0.ElementRef }, { type: ɵngcc0.Renderer2 }]; }, null); })();
