// @ks89/angular-modal-gallery/esm5/lib/modal-gallery.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DIRECTIVES } from './directives/directives';
import { COMPONENTS, ModalGalleryComponent, CarouselComponent } from './components/components';
import { KEYBOARD_CONFIGURATION, KeyboardService } from './services/keyboard.service';
import { HAMMER_GESTURE_CONFIG, HammerGestureConfig } from '@angular/platform-browser';
// to prevent bad scrolling behaviour on mobile phone with carousels.
// From @mohaxspb (https://github.com/Ks89/angular-modal-gallery/pull/187)
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './components/background/background.component';
import * as ɵngcc2 from './components/plain-gallery/plain-gallery.component';
import * as ɵngcc3 from './components/modal-gallery/modal-gallery.component';
import * as ɵngcc4 from './components/carousel/carousel.component';
import * as ɵngcc5 from './components/carousel/carousel-previews/carousel-previews.component';
import * as ɵngcc6 from './components/upper-buttons/upper-buttons.component';
import * as ɵngcc7 from './components/dots/dots.component';
import * as ɵngcc8 from './components/previews/previews.component';
import * as ɵngcc9 from './components/current-image/current-image.component';
import * as ɵngcc10 from './components/current-image/loading-spinner/loading-spinner.component';
import * as ɵngcc11 from './components/accessible.component';
import * as ɵngcc12 from './directives/click-outside.directive';
import * as ɵngcc13 from './directives/size.directive';
import * as ɵngcc14 from './directives/keyboard-navigation.directive';
import * as ɵngcc15 from './directives/wrap.directive';
import * as ɵngcc16 from './directives/direction.directive';
import * as ɵngcc17 from './directives/a-tag-bg-image.directive';
import * as ɵngcc18 from './directives/description.directive';
import * as ɵngcc19 from './directives/margin.directive';
import * as ɵngcc20 from './directives/max-size.directive';
var 
// to prevent bad scrolling behaviour on mobile phone with carousels.
// From @mohaxspb (https://github.com/Ks89/angular-modal-gallery/pull/187)
KsHammerGestureConfig = /** @class */ (function (_super) {
    tslib_1.__extends(KsHammerGestureConfig, _super);
    function KsHammerGestureConfig() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} element
     * @return {?}
     */
    KsHammerGestureConfig.prototype.buildHammer = /**
     * @param {?} element
     * @return {?}
     */
    function (element) {
        return new Hammer(element, {
            touchAction: 'pan-y'
        });
    };
KsHammerGestureConfig.ɵfac = function KsHammerGestureConfig_Factory(t) { return ɵKsHammerGestureConfig_BaseFactory(t || KsHammerGestureConfig); };
KsHammerGestureConfig.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: KsHammerGestureConfig, factory: function (t) { return KsHammerGestureConfig.ɵfac(t); } });
var ɵKsHammerGestureConfig_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(KsHammerGestureConfig);
    return KsHammerGestureConfig;
}(HammerGestureConfig));
// to prevent bad scrolling behaviour on mobile phone with carousels.
// From @mohaxspb (https://github.com/Ks89/angular-modal-gallery/pull/187)
export { KsHammerGestureConfig };
/**
 * Module to import it in the root module of your application.
 */
var GalleryModule = /** @class */ (function () {
    function GalleryModule() {
    }
    /**
     * Importing with '.forRoot()' is deprecated. I'll provide a new way to do this in next releases.
     * Will be removed in 8.0.0 or 9.0.0
     */
    /**
     * Importing with '.forRoot()' is deprecated. I'll provide a new way to do this in next releases.
     * Will be removed in 8.0.0 or 9.0.0
     * @param {?=} config
     * @return {?}
     */
    GalleryModule.forRoot = /**
     * Importing with '.forRoot()' is deprecated. I'll provide a new way to do this in next releases.
     * Will be removed in 8.0.0 or 9.0.0
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: GalleryModule,
            providers: [
                {
                    provide: KEYBOARD_CONFIGURATION,
                    useValue: config ? config : {}
                },
                {
                    provide: KeyboardService,
                    useFactory: setupKeyboardService,
                    deps: [KEYBOARD_CONFIGURATION]
                }
            ]
        };
    };
GalleryModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: GalleryModule });
GalleryModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function GalleryModule_Factory(t) { return new (t || GalleryModule)(); }, providers: [
        {
            provide: HAMMER_GESTURE_CONFIG,
            useClass: KsHammerGestureConfig
        }
    ], imports: [[CommonModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(GalleryModule, { declarations: function () { return [ɵngcc1.BackgroundComponent, ɵngcc2.PlainGalleryComponent, ɵngcc3.ModalGalleryComponent, ɵngcc4.CarouselComponent, ɵngcc5.CarouselPreviewsComponent, ɵngcc6.UpperButtonsComponent, ɵngcc7.DotsComponent, ɵngcc8.PreviewsComponent, ɵngcc9.CurrentImageComponent, ɵngcc10.LoadingSpinnerComponent, ɵngcc11.AccessibleComponent, ɵngcc12.ClickOutsideDirective, ɵngcc13.SizeDirective, ɵngcc14.KeyboardNavigationDirective, ɵngcc15.WrapDirective, ɵngcc16.DirectionDirective, ɵngcc17.ATagBgImageDirective, ɵngcc18.DescriptionDirective, ɵngcc19.MarginDirective, ɵngcc20.MaxSizeDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [ModalGalleryComponent, CarouselComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GalleryModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [COMPONENTS, DIRECTIVES],
                exports: [ModalGalleryComponent, CarouselComponent],
                providers: [
                    {
                        provide: HAMMER_GESTURE_CONFIG,
                        useClass: KsHammerGestureConfig
                    }
                ]
            }]
    }], function () { return []; }, null); })();
    return GalleryModule;
}());
export { GalleryModule };
/**
 * Function to setup the keyboard service inside the `forRoot` method.
 * @param {?} injector
 * @return {?} KeyboardService an instance of the `KeyboardService`
 */
export function setupKeyboardService(injector) {
    return new KeyboardService(injector);
}
