// @ks89/angular-modal-gallery/esm5/lib/components/accessible.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { DIRECTION_RIGHT, MOUSE_MAIN_BUTTON_CLICK, NEXT, NOTHING, PREV, ENTER_CODE, SPACE_CODE } from '../utils/user-input.util';
/**
 * Provides some useful methods to add accessibility features to subclasses.
 * In particular, it exposes a method to handle navigation event with both Keyboard and Mouse
 * and another with also the direction (right or left).
 */
import * as ɵngcc0 from '@angular/core';
var AccessibleComponent = /** @class */ (function () {
    function AccessibleComponent() {
    }
    /**
     * Method to handle navigation events with both Keyboard and Mouse.
     * @param string direction of the navigation that can be either 'next' or 'prev'
     * @param KeyboardEvent | MouseEvent event payload
     * @returns number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    /**
     * Method to handle navigation events with both Keyboard and Mouse.
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleNavigationEvent = /**
     * Method to handle navigation events with both Keyboard and Mouse.
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    function (direction, event) {
        if (!event) {
            return NOTHING;
        }
        if (event instanceof KeyboardEvent) {
            return this.handleKeyboardNavigationEvent(direction, event);
        }
        else if (event instanceof MouseEvent) {
            return this.handleMouseNavigationEvent(direction, event);
        }
        return NOTHING;
    };
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @param KeyboardEvent | MouseEvent event payload
     * @returns number 1 for NEXT and 0 for NOTHING
     */
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleImageEvent = /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    function (event) {
        if (!event) {
            return NOTHING;
        }
        if (event instanceof KeyboardEvent) {
            return this.handleImageKeyboardEvent(event);
        }
        else if (event instanceof MouseEvent) {
            return this.handleImageMouseEvent(event);
        }
        return NOTHING;
    };
    /**
     * Private method to handle keyboard events over an image.
     * @param KeyboardEvent event payload
     * @returns number 1 for NEXT and 0 for NOTHING
     */
    /**
     * Private method to handle keyboard events over an image.
     * @private
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleImageKeyboardEvent = /**
     * Private method to handle keyboard events over an image.
     * @private
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    function (event) {
        /** @type {?} */
        var key = event.code;
        if (key === SPACE_CODE || key === ENTER_CODE) {
            return NEXT;
        }
        return NOTHING;
    };
    /**
     * Private method to handle mouse events over an image.
     * @param MouseEvent event payload
     * @returns number 1 for NEXT and 0 for NOTHING
     */
    /**
     * Private method to handle mouse events over an image.
     * @private
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleImageMouseEvent = /**
     * Private method to handle mouse events over an image.
     * @private
     * @param {?} event
     * @return {?} number 1 for NEXT and 0 for NOTHING
     */
    function (event) {
        /** @type {?} */
        var mouseBtn = event.button;
        if (mouseBtn === MOUSE_MAIN_BUTTON_CLICK) {
            return NEXT;
        }
        return NOTHING;
    };
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @param string direction of the navigation that can be either 'next' or 'prev'
     * @param KeyboardEvent event payload
     * @returns number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @private
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleKeyboardNavigationEvent = /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @private
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    function (direction, event) {
        /** @type {?} */
        var key = event.code;
        if (key === SPACE_CODE || key === ENTER_CODE) {
            return direction === DIRECTION_RIGHT ? NEXT : PREV;
        }
        return NOTHING;
    };
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @param string direction of the navigation that can be either 'next' or 'prev'
     * @param MouseEvent event payload
     * @returns number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @private
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    AccessibleComponent.prototype.handleMouseNavigationEvent = /**
     * Method to handle events over an image, for instance a keypress with the Keyboard or a Mouse click.
     * @private
     * @param {?} direction
     * @param {?} event
     * @return {?} number -1 for PREV, 1 for NEXT and 0 for NOTHING
     */
    function (direction, event) {
        /** @type {?} */
        var mouseBtn = event.button;
        if (mouseBtn === MOUSE_MAIN_BUTTON_CLICK) {
            return direction === DIRECTION_RIGHT ? NEXT : PREV;
        }
        return NOTHING;
    };
    /** @nocollapse */
    AccessibleComponent.ctorParameters = function () { return []; };
AccessibleComponent.ɵfac = function AccessibleComponent_Factory(t) { return new (t || AccessibleComponent)(); };
AccessibleComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: AccessibleComponent, selectors: [["ks-accessible"]], decls: 0, vars: 0, template: function AccessibleComponent_Template(rf, ctx) { }, encapsulation: 2, changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(AccessibleComponent, [{
        type: Component,
        args: [{
                selector: 'ks-accessible',
                template: "",
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return []; }, null); })();
    return AccessibleComponent;
}());
export { AccessibleComponent };
