// ngx-joyride/esm2015/lib/directives/joyride.directive.js
import { Directive, Input, ViewContainerRef, Output, EventEmitter, Inject, PLATFORM_ID } from '@angular/core';
import { JoyrideStep } from '../models/joyride-step.class';
import { JoyrideStepsContainerService } from '../services/joyride-steps-container.service';
import { JoyrideError } from '../models/joyride-error.class';
import { Router } from '@angular/router';
import { DomRefService } from '../services/dom.service';
import { isPlatformBrowser } from '@angular/common';
import { TemplatesService } from '../services/templates.service';
import { Observable } from 'rxjs';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../services/joyride-steps-container.service';
import * as ɵngcc2 from '../services/dom.service';
import * as ɵngcc3 from '@angular/router';
import * as ɵngcc4 from '../services/templates.service';
export const NO_POSITION = 'NO_POSITION';
export class JoyrideDirective {
    constructor(joyrideStepsContainer, viewContainerRef, domService, router, templateService, platformId) {
        this.joyrideStepsContainer = joyrideStepsContainer;
        this.viewContainerRef = viewContainerRef;
        this.domService = domService;
        this.router = router;
        this.templateService = templateService;
        this.platformId = platformId;
        this.stepPosition = NO_POSITION;
        this.prev = new EventEmitter();
        this.next = new EventEmitter();
        this.done = new EventEmitter();
        this.subscriptions = [];
        this.windowRef = this.domService.getNativeWindow();
        this.step = new JoyrideStep();
    }
    ngAfterViewInit() {
        if (!isPlatformBrowser(this.platformId))
            return;
        if (this.prevTemplate)
            this.templateService.setPrevButton(this.prevTemplate);
        if (this.nextTemplate)
            this.templateService.setNextButton(this.nextTemplate);
        if (this.doneTemplate)
            this.templateService.setDoneButton(this.doneTemplate);
        if (this.counterTemplate)
            this.templateService.setCounter(this.counterTemplate);
        this.step.position = this.stepPosition;
        this.step.targetViewContainer = this.viewContainerRef;
        this.setAsyncFields(this.step);
        this.step.stepContent = this.stepContent;
        this.step.stepContentParams = this.stepContentParams;
        this.step.nextClicked = this.next;
        this.step.prevCliked = this.prev;
        this.step.tourDone = this.done;
        if (!this.name)
            throw new JoyrideError("All the steps should have the 'joyrideStep' property set with a custom name.");
        this.step.name = this.name;
        this.step.route = this.router.url.substr(0, 1) === '/' ? this.router.url.substr(1) : this.router.url;
        this.step.transformCssStyle = this.windowRef.getComputedStyle(this.viewContainerRef.element.nativeElement).transform;
        this.step.isElementOrAncestorFixed =
            this.isElementFixed(this.viewContainerRef.element) ||
                this.isAncestorsFixed(this.viewContainerRef.element.nativeElement.parentElement);
        this.joyrideStepsContainer.addStep(this.step);
    }
    ngOnChanges(changes) {
        if (changes['title'] || changes['text']) {
            this.setAsyncFields(this.step);
        }
    }
    isElementFixed(element) {
        return this.windowRef.getComputedStyle(element.nativeElement).position === 'fixed';
    }
    setAsyncFields(step) {
        if (this.title instanceof Observable) {
            this.subscriptions.push(this.title.subscribe(title => {
                step.title.next(title);
            }));
        }
        else {
            step.title.next(this.title);
        }
        if (this.text instanceof Observable) {
            this.subscriptions.push(this.text.subscribe(text => {
                step.text.next(text);
            }));
        }
        else {
            step.text.next(this.text);
        }
    }
    isAncestorsFixed(nativeElement) {
        if (!nativeElement || !nativeElement.parentElement)
            return false;
        let isElementFixed = this.windowRef.getComputedStyle(nativeElement.parentElement).position === 'fixed';
        if (nativeElement.nodeName === 'BODY') {
            return isElementFixed;
        }
        if (isElementFixed)
            return true;
        else
            return this.isAncestorsFixed(nativeElement.parentElement);
    }
    ngOnDestroy() {
        this.subscriptions.forEach(sub => {
            sub.unsubscribe();
        });
    }
}
JoyrideDirective.ɵfac = function JoyrideDirective_Factory(t) { return new (t || JoyrideDirective)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.JoyrideStepsContainerService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ViewContainerRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.DomRefService), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.Router), ɵngcc0.ɵɵdirectiveInject(ɵngcc4.TemplatesService), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID)); };
JoyrideDirective.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: JoyrideDirective, selectors: [["joyrideStep"], ["", "joyrideStep", ""]], inputs: { stepPosition: "stepPosition", name: ["joyrideStep", "name"], nextStep: "nextStep", title: "title", text: "text", stepContent: "stepContent", stepContentParams: "stepContentParams", prevTemplate: "prevTemplate", nextTemplate: "nextTemplate", doneTemplate: "doneTemplate", counterTemplate: "counterTemplate" }, outputs: { prev: "prev", next: "next", done: "done" }, features: [ɵngcc0.ɵɵNgOnChangesFeature] });
JoyrideDirective.ctorParameters = () => [
    { type: JoyrideStepsContainerService },
    { type: ViewContainerRef },
    { type: DomRefService },
    { type: Router },
    { type: TemplatesService },
    { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] }
];
JoyrideDirective.propDecorators = {
    name: [{ type: Input, args: ['joyrideStep',] }],
    nextStep: [{ type: Input }],
    title: [{ type: Input }],
    text: [{ type: Input }],
    stepPosition: [{ type: Input }],
    stepContent: [{ type: Input }],
    stepContentParams: [{ type: Input }],
    prevTemplate: [{ type: Input }],
    nextTemplate: [{ type: Input }],
    doneTemplate: [{ type: Input }],
    counterTemplate: [{ type: Input }],
    prev: [{ type: Output }],
    next: [{ type: Output }],
    done: [{ type: Output }]
};
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(JoyrideDirective, [{
        type: Directive,
        args: [{
                selector: 'joyrideStep, [joyrideStep]'
            }]
    }], function () { return [{ type: ɵngcc1.JoyrideStepsContainerService }, { type: ɵngcc0.ViewContainerRef }, { type: ɵngcc2.DomRefService }, { type: ɵngcc3.Router }, { type: ɵngcc4.TemplatesService }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }]; }, { stepPosition: [{
            type: Input
        }], prev: [{
            type: Output
        }], next: [{
            type: Output
        }], done: [{
            type: Output
        }], name: [{
            type: Input,
            args: ['joyrideStep']
        }], nextStep: [{
            type: Input
        }], title: [{
            type: Input
        }], text: [{
            type: Input
        }], stepContent: [{
            type: Input
        }], stepContentParams: [{
            type: Input
        }], prevTemplate: [{
            type: Input
        }], nextTemplate: [{
            type: Input
        }], doneTemplate: [{
            type: Input
        }], counterTemplate: [{
            type: Input
        }] }); })();
