// ngx-crystal-gallery/esm5/lib/gallery.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { LightboxComponent } from './lightbox.component';
import { LightboxMobileComponent } from './lightbox-mobile.component';
import { GalleryComponent } from './gallery.component';
import { CrystalLightbox } from './dom.service';
import { MasonryLayoutComponent } from './masonry-layout.component';
import { PinchDirective } from './pinch.directive';
import * as ɵngcc0 from '@angular/core';
var CrystalGalleryModule = /** @class */ (function () {
    function CrystalGalleryModule() {
    }
CrystalGalleryModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CrystalGalleryModule });
CrystalGalleryModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CrystalGalleryModule_Factory(t) { return new (t || CrystalGalleryModule)(); }, providers: [
        CrystalLightbox
    ], imports: [[
            CommonModule
        ]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CrystalGalleryModule, { declarations: function () { return [LightboxComponent,
        LightboxMobileComponent,
        GalleryComponent,
        MasonryLayoutComponent,
        PinchDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [GalleryComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CrystalGalleryModule, [{
        type: NgModule,
        args: [{
                declarations: [
                    LightboxComponent,
                    LightboxMobileComponent,
                    GalleryComponent,
                    MasonryLayoutComponent,
                    PinchDirective
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    GalleryComponent
                ],
                providers: [
                    CrystalLightbox
                ],
                bootstrap: [],
                entryComponents: [
                    LightboxComponent,
                    LightboxMobileComponent
                ]
            }]
    }], function () { return []; }, null); })();
    return CrystalGalleryModule;
}());
export { CrystalGalleryModule };
