// ngx-crystal-gallery/esm5/lib/gallery.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef, ViewChild } from '@angular/core';
import { CrystalLightbox } from './dom.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './dom.service';
import * as ɵngcc2 from '@angular/common';
import * as ɵngcc3 from './masonry-layout.component';

var _c0 = ["contentWrapper"];
function GalleryComponent_masonry_layout_3_Template(rf, ctx) { if (rf & 1) {
    var _r3 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "masonry-layout", 3);
    ɵngcc0.ɵɵlistener("events", function GalleryComponent_masonry_layout_3_Template_masonry_layout_events_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r3); var ctx_r2 = ɵngcc0.ɵɵnextContext(); return ctx_r2.handleMasonryLayoutEvents($event); });
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("max-height", ctx_r1.masonryMaxHeight)("gutter", ctx_r1.masonryGutter)("masonry", ctx_r1.masonryState)("class-name", "cg-image")("images", ctx_r1.images);
} }
var _c1 = ["*"];
var GalleryComponent = /** @class */ (function () {
    function GalleryComponent(lightbox, elementRef) {
        this.lightbox = lightbox;
        this.elementRef = elementRef;
        this.config = {};
    }
    Object.defineProperty(GalleryComponent.prototype, "isNgContent", {
        get: /**
         * @return {?}
         */
        function () {
            return this.contentWrapper.nativeElement.children.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GalleryComponent.prototype, "masonryState", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.masonry;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GalleryComponent.prototype, "masonryMaxHeight", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.masonryMaxHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(GalleryComponent.prototype, "masonryGutter", {
        get: /**
         * @return {?}
         */
        function () {
            return this.config.masonryGutter;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} images
     * @param {?} i
     * @return {?}
     */
    GalleryComponent.prototype.showLightbox = /**
     * @param {?} images
     * @param {?} i
     * @return {?}
     */
    function (images, i) {
        this.config.index = i;
        this.lightbox.open(images, this.config);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    GalleryComponent.prototype.handleMasonryLayoutEvents = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showLightbox(this.images, event.index);
    };
    /** @nocollapse */
    GalleryComponent.ctorParameters = function () { return [
        { type: CrystalLightbox },
        { type: ElementRef }
    ]; };
    GalleryComponent.propDecorators = {
        images: [{ type: Input, args: ['images',] }],
        config: [{ type: Input, args: ['config',] }],
        contentWrapper: [{ type: ViewChild, args: ['contentWrapper',] }]
    };
GalleryComponent.ɵfac = function GalleryComponent_Factory(t) { return new (t || GalleryComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.CrystalLightbox), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ElementRef)); };
GalleryComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: GalleryComponent, selectors: [["crystal-gallery"]], viewQuery: function GalleryComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(_c0, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.contentWrapper = _t.first);
    } }, inputs: { config: "config", images: "images" }, ngContentSelectors: _c1, decls: 4, vars: 2, consts: [[3, "hidden"], ["contentWrapper", ""], ["class", "cg-wrapper", 3, "max-height", "gutter", "masonry", "class-name", "images", "events", 4, "ngIf"], [1, "cg-wrapper", 3, "max-height", "gutter", "masonry", "class-name", "images", "events"]], template: function GalleryComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef();
        ɵngcc0.ɵɵelementStart(0, "div", 0, 1);
        ɵngcc0.ɵɵprojection(2);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(3, GalleryComponent_masonry_layout_3_Template, 1, 5, "masonry-layout", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("hidden", ctx.isNgContent);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isNgContent);
    } }, directives: [ɵngcc2.NgIf, ɵngcc3.MasonryLayoutComponent], styles: ["[_nghost-%COMP%]{display:block;width:100%}[_nghost-%COMP%]   .cg-wrapper[_ngcontent-%COMP%]{display:block;width:100%;overflow:hidden}[_nghost-%COMP%]   .cg-wrapper.cg-show[_ngcontent-%COMP%]{height:auto;position:relative;display:table}[_nghost-%COMP%]   .cg-image[_ngcontent-%COMP%]   img[_ngcontent-%COMP%]{cursor:pointer;width:100%}"] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(GalleryComponent, [{
        type: Component,
        args: [{
                selector: 'crystal-gallery',
                template: "<div #contentWrapper [hidden]=\"isNgContent\">\n    <ng-content></ng-content>\n</div>\n\n<masonry-layout \n    class=\"cg-wrapper\"\n    *ngIf=\"!isNgContent\" \n\n    [max-height]=\"masonryMaxHeight\" \n    [gutter]=\"masonryGutter\"\n    [masonry]=\"masonryState\"\n    [class-name]=\"'cg-image'\"\n    [images]=\"images\"\n    (events)=\"handleMasonryLayoutEvents($event)\"></masonry-layout>",
                styles: [":host{display:block;width:100%}:host .cg-wrapper{display:block;width:100%;overflow:hidden}:host .cg-wrapper.cg-show{height:auto;position:relative;display:table}:host .cg-image img{cursor:pointer;width:100%}"]
            }]
    }], function () { return [{ type: ɵngcc1.CrystalLightbox }, { type: ɵngcc0.ElementRef }]; }, { config: [{
            type: Input,
            args: ['config']
        }], images: [{
            type: Input,
            args: ['images']
        }], contentWrapper: [{
            type: ViewChild,
            args: ['contentWrapper']
        }] }); })();
    return GalleryComponent;
}());
export { GalleryComponent };
if (false) {
    /** @type {?} */
    GalleryComponent.prototype.images;
    /** @type {?} */
    GalleryComponent.prototype.config;
    /** @type {?} */
    GalleryComponent.prototype.contentWrapper;
    /** @type {?} */
    GalleryComponent.prototype.lightbox;
    /** @type {?} */
    GalleryComponent.prototype.elementRef;
}
