// ngx-dropzone-wrapper/esm2015/lib/dropzone.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { DropzoneComponent } from './dropzone.component';
import { DropzoneDirective } from './dropzone.directive';
import * as ɵngcc0 from '@angular/core';
let DropzoneModule = class DropzoneModule {
};
DropzoneModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: DropzoneModule });
DropzoneModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function DropzoneModule_Factory(t) { return new (t || DropzoneModule)(); }, imports: [[CommonModule], CommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(DropzoneModule, { declarations: function () { return [DropzoneComponent, DropzoneDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CommonModule, DropzoneComponent, DropzoneDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(DropzoneModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [DropzoneComponent, DropzoneDirective],
                exports: [CommonModule, DropzoneComponent, DropzoneDirective]
            }]
    }], null, null); })();
export { DropzoneModule };
