// @ks89/angular-modal-gallery/esm5/lib/components/background/background.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
/**
 * Component with the semi-transparent background.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

function BackgroundComponent_div_0_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "div", 1);
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("title", ctx_r0.accessibilityConfig == null ? null : ctx_r0.accessibilityConfig.backgroundTitle);
    ɵngcc0.ɵɵattribute("aria-label", ctx_r0.accessibilityConfig == null ? null : ctx_r0.accessibilityConfig.backgroundAriaLabel);
} }
var BackgroundComponent = /** @class */ (function () {
    function BackgroundComponent() {
    }
    BackgroundComponent.propDecorators = {
        isOpen: [{ type: Input }],
        accessibilityConfig: [{ type: Input }]
    };
BackgroundComponent.ɵfac = function BackgroundComponent_Factory(t) { return new (t || BackgroundComponent)(); };
BackgroundComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: BackgroundComponent, selectors: [["ks-background"]], inputs: { isOpen: "isOpen", accessibilityConfig: "accessibilityConfig" }, decls: 1, vars: 1, consts: [["class", "ng-overlay", 3, "title", 4, "ngIf"], [1, "ng-overlay", 3, "title"]], template: function BackgroundComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵtemplate(0, BackgroundComponent_div_0_Template, 1, 2, "div", 0);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("ngIf", ctx.isOpen);
    } }, directives: [ɵngcc1.NgIf], styles: [".ng-overlay[_ngcontent-%COMP%]{position:fixed;top:0;left:0;width:100vw;height:100vh;background:#000;opacity:.8;z-index:9999}"], changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(BackgroundComponent, [{
        type: Component,
        args: [{
                selector: 'ks-background',
                template: "<div class=\"ng-overlay\" *ngIf=\"isOpen\"\n     [attr.aria-label]=\"accessibilityConfig?.backgroundAriaLabel\"\n     [title]=\"accessibilityConfig?.backgroundTitle\"></div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: [".ng-overlay{position:fixed;top:0;left:0;width:100vw;height:100vh;background:#000;opacity:.8;z-index:9999}"]
            }]
    }], function () { return []; }, { isOpen: [{
            type: Input
        }], accessibilityConfig: [{
            type: Input
        }] }); })();
    return BackgroundComponent;
}());
export { BackgroundComponent };
if (false) {
    /**
     * Boolean that it is true if the modal gallery is visible,
     * so also this component should be visible.
     * @type {?}
     */
    BackgroundComponent.prototype.isOpen;
    /**
     * Object of type `AccessibilityConfig` to init custom accessibility features.
     * For instance, it contains titles, alt texts, aria-labels and so on.
     * @type {?}
     */
    BackgroundComponent.prototype.accessibilityConfig;
}
