// @ks89/angular-modal-gallery/esm5/lib/components/modal-gallery/modal-gallery.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 The MIT License (MIT)

 Copyright (c) 2017-2018 Stefano Cappa (Ks89)

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, HostListener, Inject, Input, Output, PLATFORM_ID, ViewChild } from '@angular/core';
import { isPlatformBrowser, isPlatformServer } from '@angular/common';
import { ImageModalEvent } from '../../model/image.class';
import { Action } from '../../model/action.enum';
import { KeyboardService } from '../../services/keyboard.service';
import { GalleryService } from '../../services/gallery.service';
import { CurrentImageComponent } from '../current-image/current-image.component';
import { AdvancedLayout } from '../../model/plain-gallery-config.interface';
import { KS_DEFAULT_ACCESSIBILITY_CONFIG } from '../accessibility-default';
import { getIndex } from '../../utils/image.util';
import { IdValidatorService } from '../../services/id-validator.service';
/**
 * Main Component of this library with both the plain and modal galleries.
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../services/keyboard.service';
import * as ɵngcc2 from '../../services/gallery.service';
import * as ɵngcc3 from '../../services/id-validator.service';
import * as ɵngcc4 from '../plain-gallery/plain-gallery.component';
import * as ɵngcc5 from '../background/background.component';
import * as ɵngcc6 from '@angular/common';
import * as ɵngcc7 from '../../directives/click-outside.directive';
import * as ɵngcc8 from '../upper-buttons/upper-buttons.component';
import * as ɵngcc9 from '../current-image/current-image.component';
import * as ɵngcc10 from '../dots/dots.component';
import * as ɵngcc11 from '../previews/previews.component';

function ModalGalleryComponent_div_2_Template(rf, ctx) { if (rf & 1) {
    var _r2 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "div", 3);
    ɵngcc0.ɵɵlistener("clickOutside", function ModalGalleryComponent_div_2_Template_div_clickOutside_0_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r1 = ɵngcc0.ɵɵnextContext(); return ctx_r1.onClickOutside($event); });
    ɵngcc0.ɵɵelementStart(1, "div", 4);
    ɵngcc0.ɵɵelementStart(2, "div", 5);
    ɵngcc0.ɵɵelementStart(3, "ks-upper-buttons", 6);
    ɵngcc0.ɵɵlistener("delete", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_delete_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r3 = ɵngcc0.ɵɵnextContext(); return ctx_r3.onDelete($event); })("navigate", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_navigate_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.onNavigate($event); })("download", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_download_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r5 = ɵngcc0.ɵɵnextContext(); return ctx_r5.onDownload($event); })("close", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_close_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r6 = ɵngcc0.ɵɵnextContext(); return ctx_r6.onCloseGallery($event); })("fullscreen", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_fullscreen_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r7 = ɵngcc0.ɵɵnextContext(); return ctx_r7.onFullScreen($event); })("customEmit", function ModalGalleryComponent_div_2_Template_ks_upper_buttons_customEmit_3_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r8 = ɵngcc0.ɵɵnextContext(); return ctx_r8.onCustomEmit($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(4, "ks-current-image", 7);
    ɵngcc0.ɵɵlistener("loadImage", function ModalGalleryComponent_div_2_Template_ks_current_image_loadImage_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r9 = ɵngcc0.ɵɵnextContext(); return ctx_r9.onImageLoad($event); })("changeImage", function ModalGalleryComponent_div_2_Template_ks_current_image_changeImage_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r10 = ɵngcc0.ɵɵnextContext(); return ctx_r10.onChangeCurrentImage($event); })("close", function ModalGalleryComponent_div_2_Template_ks_current_image_close_4_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r11 = ɵngcc0.ɵɵnextContext(); return ctx_r11.onCloseGallery($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "div");
    ɵngcc0.ɵɵelementStart(6, "ks-dots", 8);
    ɵngcc0.ɵɵlistener("clickDot", function ModalGalleryComponent_div_2_Template_ks_dots_clickDot_6_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r12 = ɵngcc0.ɵɵnextContext(); return ctx_r12.onClickDot($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(7, "ks-previews", 9);
    ɵngcc0.ɵɵlistener("clickPreview", function ModalGalleryComponent_div_2_Template_ks_previews_clickPreview_7_listener($event) { ɵngcc0.ɵɵrestoreView(_r2); var ctx_r13 = ɵngcc0.ɵɵnextContext(); return ctx_r13.onClickPreview($event); });
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    var ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("title", ctx_r0.accessibilityConfig.modalGalleryContentTitle)("clickOutsideEnable", ctx_r0.enableCloseOutside);
    ɵngcc0.ɵɵattribute("aria-label", ctx_r0.accessibilityConfig.modalGalleryContentAriaLabel);
    ɵngcc0.ɵɵadvance(3);
    ɵngcc0.ɵɵproperty("currentImage", ctx_r0.currentImage)("buttonsConfig", ctx_r0.buttonsConfig);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("id", ctx_r0.id)("images", ctx_r0.images)("currentImage", ctx_r0.currentImage)("isOpen", ctx_r0.opened)("slideConfig", ctx_r0.configSlide)("keyboardConfig", ctx_r0.keyboardConfig)("accessibilityConfig", ctx_r0.accessibilityConfig)("currentImageConfig", ctx_r0.currentImageConfig);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("images", ctx_r0.images)("currentImage", ctx_r0.currentImage)("dotsConfig", ctx_r0.dotsConfig)("accessibilityConfig", ctx_r0.accessibilityConfig);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("images", ctx_r0.images)("currentImage", ctx_r0.currentImage)("previewConfig", ctx_r0.previewConfig)("accessibilityConfig", ctx_r0.accessibilityConfig);
} }
var ModalGalleryComponent = /** @class */ (function () {
    /**
     * Constructor with the injection of ´KeyboardService´, an object to support Server-Side Rendering and other useful services.
     */
    function ModalGalleryComponent(keyboardService, galleryService, platformId, changeDetectorRef, idValidatorService) {
        this.keyboardService = keyboardService;
        this.galleryService = galleryService;
        this.platformId = platformId;
        this.changeDetectorRef = changeDetectorRef;
        this.idValidatorService = idValidatorService;
        /**
         * Boolean to enable modal-gallery close behaviour when clicking
         * on the semi-transparent background. Enabled by default.
         */
        this.enableCloseOutside = true;
        /**
         * Object of type `AccessibilityConfig` to init custom accessibility features.
         * For instance, it contains titles, alt texts, aria-labels and so on.
         */
        this.accessibilityConfig = KS_DEFAULT_ACCESSIBILITY_CONFIG;
        /**
         * Output to emit an event when the modal gallery is closed.
         */
        this.close = new EventEmitter();
        /**
         * Output to emit an event when an image is changed.
         */
        this.show = new EventEmitter();
        /**
         * Output to emit an event when the current image is the first one.
         */
        this.firstImage = new EventEmitter();
        /**
         * Output to emit an event when the current image is the last one.
         */
        this.lastImage = new EventEmitter();
        /**
         * Output to emit an event when the modal gallery is closed.
         */
        this.hasData = new EventEmitter();
        /**
         * Output to emit an event when a button is clicked, but before that the action is triggered.
         */
        this.buttonBeforeHook = new EventEmitter();
        /**
         * Output to emit an event when a button is clicked, but after that the action is triggered.
         */
        this.buttonAfterHook = new EventEmitter();
        /**
         * Output to emit an event when someone clicks either an arrow of modal gallery or also in previews.
         */
        this.arrow = new EventEmitter();
        /**
         * Boolean that it is true if the modal gallery is visible. False by default.
         */
        this.opened = false;
        /**
         * Boolean to open the modal gallery. False by default.
         */
        this.showGallery = false;
    }
    /**
     * HostListener to catch browser's back button and destroy the gallery.
     * This prevents weired behaviour about scrolling.
     * Added to fix this issue: https://github.com/Ks89/angular-modal-gallery/issues/159
     */
    /**
     * HostListener to catch browser's back button and destroy the gallery.
     * This prevents weired behaviour about scrolling.
     * Added to fix this issue: https://github.com/Ks89/angular-modal-gallery/issues/159
     * @param {?} e
     * @return {?}
     */
    ModalGalleryComponent.prototype.onPopState = /**
     * HostListener to catch browser's back button and destroy the gallery.
     * This prevents weired behaviour about scrolling.
     * Added to fix this issue: https://github.com/Ks89/angular-modal-gallery/issues/159
     * @param {?} e
     * @return {?}
     */
    function (e) {
        this.closeGallery();
    };
    /**
     * Method ´ngOnChanges´ to re-init images if input is changed.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called before `ngOnInit()` and whenever one or more data-bound input properties change.
     * @param changes `SimpleChanges` object of current and previous property values provided by Angular.
     */
    /**
     * Method ´ngOnChanges´ to re-init images if input is changed.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called before `ngOnInit()` and whenever one or more data-bound input properties change.
     * @param {?} changes `SimpleChanges` object of current and previous property values provided by Angular.
     * @return {?}
     */
    ModalGalleryComponent.prototype.ngOnChanges = /**
     * Method ´ngOnChanges´ to re-init images if input is changed.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called before `ngOnInit()` and whenever one or more data-bound input properties change.
     * @param {?} changes `SimpleChanges` object of current and previous property values provided by Angular.
     * @return {?}
     */
    function (changes) {
        /** @type {?} */
        var imagesChange = changes.modalImages;
        /** @type {?} */
        var plainGalleryConfigChange = changes.plainGalleryConfig;
        if (imagesChange && !imagesChange.firstChange && imagesChange.previousValue !== imagesChange.currentValue) {
            this.initImages();
        }
        if (plainGalleryConfigChange) {
            // const prevPlainGalleryConfigChange: any = plainGalleryConfigChange.previousValue;
            /** @type {?} */
            var currPlainGalleryConfigChange = plainGalleryConfigChange.currentValue;
            if (currPlainGalleryConfigChange.layout &&
                currPlainGalleryConfigChange.layout instanceof AdvancedLayout &&
                currPlainGalleryConfigChange.layout.modalOpenerByIndex !== -1) {
                // console.log('opening modal gallery from custom plain gallery, index: ', currPlainGalleryConfigChange);
                this.showModalGallery(currPlainGalleryConfigChange.layout.modalOpenerByIndex);
            }
        }
    };
    /**
     * Method ´ngOnInit´ to init images calling `initImages()`.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     */
    /**
     * Method ´ngOnInit´ to init images calling `initImages()`.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    ModalGalleryComponent.prototype.ngOnInit = /**
     * Method ´ngOnInit´ to init images calling `initImages()`.
     * This is an Angular's lifecycle hook, so its called automatically by Angular itself.
     * In particular, it's called only one time!!!
     * @return {?}
     */
    function () {
        var _this = this;
        this.idValidatorService.checkAndAdd(this.id);
        // id is a mandatory input and must a number > 0
        if ((!this.id && this.id !== 0) || this.id < 0) {
            throw new Error("'[id]=\"a number >= 0\"' is a mandatory input from 6.0.0 in angular-modal-gallery." +
                "If you are using multiple instances of this library, please be sure to use different ids");
        }
        // call initImages to init images and to emit `hasData` event
        this.initImages();
        /** @type {?} */
        var defaultSlideConfig = {
            infinite: false,
            playConfig: (/** @type {?} */ ({ autoPlay: false, interval: 5000, pauseOnHover: true })),
            sidePreviews: (/** @type {?} */ ({ show: true, size: { width: '100px', height: 'auto' } }))
        };
        this.configSlide = Object.assign({}, defaultSlideConfig, this.slideConfig);
        this.galleryServiceNavigateSubscription = this.galleryService.navigate.subscribe((/**
         * @param {?} payload
         * @return {?}
         */
        function (payload) {
            if (!payload) {
                return;
            }
            // if galleryId is not valid OR galleryId is related to another instance and not this one
            if (payload.galleryId === undefined || payload.galleryId < 0 || payload.galleryId !== _this.id) {
                return;
            }
            // if image index is not valid
            if (payload.index < 0 || payload.index > _this.images.length) {
                return;
            }
            _this.showModalGallery(payload.index, true);
        }));
        this.galleryServiceCloseSubscription = this.galleryService.close.subscribe((/**
         * @param {?} galleryId
         * @return {?}
         */
        function (galleryId) {
            if (galleryId < 0 || _this.id !== galleryId) {
                return;
            }
            _this.closeGallery(Action.NORMAL, true);
        }));
        this.galleryServiceUpdateSubscription = this.galleryService.update.subscribe((/**
         * @param {?} payload
         * @return {?}
         */
        function (payload) {
            if (!payload) {
                return;
            }
            // if galleryId is not valid OR galleryId is related to another instance and not this one
            if (payload.galleryId === undefined || payload.galleryId < 0 || payload.galleryId !== _this.id) {
                return;
            }
            // if either image index or image are not valid
            if (payload.index < 0 || payload.index > _this.images.length || !payload.image) {
                return;
            }
            /** @type {?} */
            var currentIndex = getIndex(payload.image, _this.images);
            _this.images = _this.images.map((/**
             * @param {?} image
             * @param {?} index
             * @return {?}
             */
            function (image, index) {
                if (image.id === payload.index) {
                    return (/** @type {?} */ (payload.image));
                }
                return image;
            }));
            if (currentIndex === payload.index) {
                _this.currentImage = _this.images[payload.index];
            }
            _this.changeDetectorRef.markForCheck();
        }));
        this.galleryServiceAutoPlaySubscription = this.galleryService.autoPlay.subscribe((/**
         * @param {?} payload
         * @return {?}
         */
        function (payload) {
            // if galleryId is not valid OR galleryId is related to another instance and not this one
            if (payload.galleryId === undefined || payload.galleryId < 0 || payload.galleryId !== _this.id) {
                return;
            }
            _this.configSlide.playConfig.autoPlay = payload.result;
        }));
    };
    /**
     * Method called by custom upper buttons.
     * @param ButtonEvent event payload
     */
    /**
     * Method called by custom upper buttons.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onCustomEmit = /**
     * Method called by custom upper buttons.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        // console.log('on onCustomEmit', eventToEmit);
        this.buttonAfterHook.emit(eventToEmit);
    };
    // TODO implement on refresh
    // /**
    //  * Method called by the refresh upper button.
    //  * STILL NOT IMPLEMENTED, SO DON'T USE IT
    //  * @param ButtonEvent event payload
    //  */
    // onRefresh(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //
    //   this.buttonBeforeHook.emit(eventToEmit);
    //   // console.log('TODO implement on refresh inside the library', eventToEmit);
    //
    //   this.currentImage = Object.assign({}, this.currentImage, { previouslyLoaded: false });
    //
    //   // TODO add logic to hide and show the current image
    //
    //   // console.log('onRefresh', this.currentImage);
    //
    //   // const indexNum: number = this.currentImageComponent.getIndex();
    //
    //   // this.images = this.images.map((val: InternalLibImage, index: number) => {
    //   //   if (index !== 2) {
    //   //     return val;
    //   //   } else {
    //   //     const img: InternalLibImage = Object.assign({}, val, {previouslyLoaded: false});
    //   //     return img;
    //   //   }
    //   // });
    //   //
    //   // this.closeGallery();
    //   // this.showModalGallery(2);
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    // /**
    //  * Method called by the rotate upper button.
    //  * @param ButtonEvent event payload
    //  */
    // onRotate(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //   this.buttonBeforeHook.emit(eventToEmit);
    //
    //   // TODO implement rotation logic
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    /**
     * Method called by the full-screen upper button.
     * @param ButtonEvent event payload
     */
    // TODO implement on refresh
    // /**
    //  * Method called by the refresh upper button.
    //  * STILL NOT IMPLEMENTED, SO DON'T USE IT
    //  * @param ButtonEvent event payload
    //  */
    // onRefresh(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //
    //   this.buttonBeforeHook.emit(eventToEmit);
    //   // console.log('TODO implement on refresh inside the library', eventToEmit);
    //
    //   this.currentImage = Object.assign({}, this.currentImage, { previouslyLoaded: false });
    //
    //   // TODO add logic to hide and show the current image
    //
    //   // console.log('onRefresh', this.currentImage);
    //
    //   // const indexNum: number = this.currentImageComponent.getIndex();
    //
    //   // this.images = this.images.map((val: InternalLibImage, index: number) => {
    //   //   if (index !== 2) {
    //   //     return val;
    //   //   } else {
    //   //     const img: InternalLibImage = Object.assign({}, val, {previouslyLoaded: false});
    //   //     return img;
    //   //   }
    //   // });
    //   //
    //   // this.closeGallery();
    //   // this.showModalGallery(2);
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    // /**
    //  * Method called by the rotate upper button.
    //  * @param ButtonEvent event payload
    //  */
    // onRotate(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //   this.buttonBeforeHook.emit(eventToEmit);
    //
    //   // TODO implement rotation logic
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    /**
     * Method called by the full-screen upper button.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onFullScreen = 
    // TODO implement on refresh
    // /**
    //  * Method called by the refresh upper button.
    //  * STILL NOT IMPLEMENTED, SO DON'T USE IT
    //  * @param ButtonEvent event payload
    //  */
    // onRefresh(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //
    //   this.buttonBeforeHook.emit(eventToEmit);
    //   // console.log('TODO implement on refresh inside the library', eventToEmit);
    //
    //   this.currentImage = Object.assign({}, this.currentImage, { previouslyLoaded: false });
    //
    //   // TODO add logic to hide and show the current image
    //
    //   // console.log('onRefresh', this.currentImage);
    //
    //   // const indexNum: number = this.currentImageComponent.getIndex();
    //
    //   // this.images = this.images.map((val: InternalLibImage, index: number) => {
    //   //   if (index !== 2) {
    //   //     return val;
    //   //   } else {
    //   //     const img: InternalLibImage = Object.assign({}, val, {previouslyLoaded: false});
    //   //     return img;
    //   //   }
    //   // });
    //   //
    //   // this.closeGallery();
    //   // this.showModalGallery(2);
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    // /**
    //  * Method called by the rotate upper button.
    //  * @param ButtonEvent event payload
    //  */
    // onRotate(event: ButtonEvent) {
    //   const eventToEmit: ButtonEvent = this.getButtonEventToEmit(event);
    //   this.buttonBeforeHook.emit(eventToEmit);
    //
    //   // TODO implement rotation logic
    //
    //   this.buttonAfterHook.emit(eventToEmit);
    // }
    /**
     * Method called by the full-screen upper button.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        /** @type {?} */
        var doc = (/** @type {?} */ (document));
        /** @type {?} */
        var docEl = (/** @type {?} */ (document.documentElement));
        /** @type {?} */
        var fullscreenDisabled = !doc.fullscreenElement && !doc.webkitFullscreenElement && !doc.mozFullScreenElement && !doc.msFullscreenElement;
        if (fullscreenDisabled) {
            if (docEl.requestFullscreen) {
                docEl.requestFullscreen();
            }
            else if (docEl.webkitRequestFullscreen) {
                docEl.webkitRequestFullscreen();
            }
            else if (docEl.mozRequestFullScreen) {
                docEl.mozRequestFullScreen();
            }
            else if (docEl.msRequestFullscreen) {
                docEl.msRequestFullscreen();
            }
        }
        else {
            if (doc.exitFullscreen) {
                doc.exitFullscreen();
            }
            else if (doc.msExitFullscreen) {
                doc.msExitFullscreen();
            }
            else if (doc.mozCancelFullScreen) {
                doc.mozCancelFullScreen();
            }
            else if (doc.webkitExitFullscreen) {
                doc.webkitExitFullscreen();
            }
        }
        this.buttonAfterHook.emit(eventToEmit);
    };
    /**
     * Method called by the delete upper button.
     * @param ButtonEvent event payload
     */
    /**
     * Method called by the delete upper button.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onDelete = /**
     * Method called by the delete upper button.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        if (this.images.length === 1) {
            this.closeGallery();
        }
        /** @type {?} */
        var imageIndexToDelete = this.currentImageComponent.getIndexToDelete(event.image);
        if (imageIndexToDelete === this.images.length - 1) {
            // last image
            this.currentImageComponent.prevImage();
        }
        else {
            this.currentImageComponent.nextImage();
        }
        this.buttonAfterHook.emit(eventToEmit);
    };
    /**
     * Method called by the navigate upper button.
     * @param ButtonEvent event payload
     */
    /**
     * Method called by the navigate upper button.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onNavigate = /**
     * Method called by the navigate upper button.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        // To support SSR
        if (isPlatformBrowser(this.platformId)) {
            if (eventToEmit.image && eventToEmit.image.modal.extUrl) {
                // where I should open this link? The current tab or another one?
                if (eventToEmit.button && eventToEmit.button.extUrlInNewTab) {
                    // in this case I should use target _blank to open the url in a new tab, however these is a security issue.
                    // Prevent Reverse Tabnabbing's attacks (https://www.owasp.org/index.php/Reverse_Tabnabbing)
                    // Some resources:
                    // - https://www.owasp.org/index.php/HTML5_Security_Cheat_Sheet#Tabnabbing
                    // - https://medium.com/@jitbit/target-blank-the-most-underestimated-vulnerability-ever-96e328301f4c
                    // - https://developer.mozilla.org/en-US/docs/Web/API/Window/open
                    /** @type {?} */
                    var newWindow = window.open(eventToEmit.image.modal.extUrl, 'noopener,noreferrer,');
                    newWindow.opener = null; // required to prevent security issues
                }
                else {
                    window.location.href = eventToEmit.image.modal.extUrl;
                }
            }
        }
        this.buttonAfterHook.emit(eventToEmit);
    };
    /**
     * Method called by the download upper button.
     * @param ButtonEvent event payload
     */
    /**
     * Method called by the download upper button.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onDownload = /**
     * Method called by the download upper button.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        this.downloadImage();
        this.buttonAfterHook.emit(eventToEmit);
    };
    /**
     * Method called by the close upper button.
     * @param ButtonEvent event payload
     * @param Action action that triggered the close method. `Action.NORMAL` by default
     */
    /**
     * Method called by the close upper button.
     * @param {?} event
     * @param {?=} action
     * @return {?}
     */
    ModalGalleryComponent.prototype.onCloseGallery = /**
     * Method called by the close upper button.
     * @param {?} event
     * @param {?=} action
     * @return {?}
     */
    function (event, action) {
        if (action === void 0) { action = Action.NORMAL; }
        /** @type {?} */
        var eventToEmit = this.getButtonEventToEmit(event);
        this.buttonBeforeHook.emit(eventToEmit);
        this.closeGallery(action);
        this.buttonAfterHook.emit(eventToEmit);
    };
    /**
     * Method to close the modal gallery specifying the action.
     * It also reset the `keyboardService` to prevent multiple listeners.
     * @param Action action type. `Action.NORMAL` by default
     * @param boolean isCalledByService is true if called by gallery.service, otherwise false
     */
    /**
     * Method to close the modal gallery specifying the action.
     * It also reset the `keyboardService` to prevent multiple listeners.
     * @param {?=} action
     * @param {?=} isCalledByService
     * @return {?}
     */
    ModalGalleryComponent.prototype.closeGallery = /**
     * Method to close the modal gallery specifying the action.
     * It also reset the `keyboardService` to prevent multiple listeners.
     * @param {?=} action
     * @param {?=} isCalledByService
     * @return {?}
     */
    function (action, isCalledByService) {
        if (action === void 0) { action = Action.NORMAL; }
        if (isCalledByService === void 0) { isCalledByService = false; }
        this.close.emit(new ImageModalEvent(action, true));
        this.opened = false;
        this.keyboardService.reset();
        // shows scrollbar
        document.body.style.overflow = 'visible';
        if (isCalledByService) {
            // the following is required, otherwise the view will not be updated
            // this happens only if called by gallery.service
            this.changeDetectorRef.markForCheck();
        }
    };
    /**
     * Method called when you click on an image of your plain (or inline) gallery.
     * @param number index of the clicked image
     */
    /**
     * Method called when you click on an image of your plain (or inline) gallery.
     * @param {?} index
     * @return {?}
     */
    ModalGalleryComponent.prototype.onShowModalGallery = /**
     * Method called when you click on an image of your plain (or inline) gallery.
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.showModalGallery(index);
    };
    /**
     * Method to show the modal gallery displaying the image with
     * the index specified as input parameter.
     * It will also register a new `keyboardService` to catch keyboard's events to download the current
     * image with keyboard's shortcuts. This service, will be removed either when modal gallery component
     * will be destroyed or when the gallery is closed invoking the `closeGallery` method.
     * @param number index of the image to show
     * @param boolean isCalledByService is true if called by gallery.service, otherwise false
     */
    /**
     * Method to show the modal gallery displaying the image with
     * the index specified as input parameter.
     * It will also register a new `keyboardService` to catch keyboard's events to download the current
     * image with keyboard's shortcuts. This service, will be removed either when modal gallery component
     * will be destroyed or when the gallery is closed invoking the `closeGallery` method.
     * @param {?} index
     * @param {?=} isCalledByService
     * @return {?}
     */
    ModalGalleryComponent.prototype.showModalGallery = /**
     * Method to show the modal gallery displaying the image with
     * the index specified as input parameter.
     * It will also register a new `keyboardService` to catch keyboard's events to download the current
     * image with keyboard's shortcuts. This service, will be removed either when modal gallery component
     * will be destroyed or when the gallery is closed invoking the `closeGallery` method.
     * @param {?} index
     * @param {?=} isCalledByService
     * @return {?}
     */
    function (index, isCalledByService) {
        var _this = this;
        if (isCalledByService === void 0) { isCalledByService = false; }
        // hides scrollbar
        document.body.style.overflow = 'hidden';
        this.keyboardService.add((/**
         * @param {?} event
         * @param {?} combo
         * @return {?}
         */
        function (event, combo) {
            if (event.preventDefault) {
                event.preventDefault();
            }
            else {
                // internet explorer
                event.returnValue = false;
            }
            _this.downloadImage();
        }));
        this.opened = true;
        this.currentImage = this.images[index];
        // emit a new ImageModalEvent with the index of the current image
        this.show.emit(new ImageModalEvent(Action.LOAD, index + 1));
        if (isCalledByService) {
            // the following is required, otherwise the view will not be updated
            // this happens only if called by gallery.service
            this.changeDetectorRef.markForCheck();
        }
    };
    /**
     * Method called when the image changes and used to update the `currentImage` object.
     * @param ImageModalEvent event payload
     */
    /**
     * Method called when the image changes and used to update the `currentImage` object.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onChangeCurrentImage = /**
     * Method called when the image changes and used to update the `currentImage` object.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        /** @type {?} */
        var newIndex = (/** @type {?} */ (event.result));
        if (newIndex < 0 || newIndex >= this.images.length) {
            return;
        }
        this.currentImage = this.images[newIndex];
        // emit first/last event based on newIndex value
        this.emitBoundaryEvent(event.action, newIndex);
        // emit current visible image index
        this.show.emit(new ImageModalEvent(event.action, newIndex + 1));
    };
    /**
     * @return {?}
     */
    ModalGalleryComponent.prototype.isPlainGalleryVisible = /**
     * @return {?}
     */
    function () {
        if (this.plainGalleryConfig && this.plainGalleryConfig.layout && this.plainGalleryConfig.layout instanceof AdvancedLayout) {
            return !this.plainGalleryConfig.layout.hideDefaultPlainGallery;
        }
        return true;
    };
    /**
     * Method called when you click 'outside' (i.e. on the semi-transparent background)
     * to close the modal gallery if `enableCloseOutside` is true.
     * @param boolean event payload. True to close the modal gallery, false otherwise
     */
    /**
     * Method called when you click 'outside' (i.e. on the semi-transparent background)
     * to close the modal gallery if `enableCloseOutside` is true.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onClickOutside = /**
     * Method called when you click 'outside' (i.e. on the semi-transparent background)
     * to close the modal gallery if `enableCloseOutside` is true.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (event && this.enableCloseOutside) {
            this.closeGallery(Action.CLICK);
        }
    };
    /**
     * Method called when an image is loaded and the loading spinner has gone.
     * It sets the previouslyLoaded flag inside the Image to hide loading spinner when displayed again.
     * @param ImageLoadEvent event payload
     */
    /**
     * Method called when an image is loaded and the loading spinner has gone.
     * It sets the previouslyLoaded flag inside the Image to hide loading spinner when displayed again.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onImageLoad = /**
     * Method called when an image is loaded and the loading spinner has gone.
     * It sets the previouslyLoaded flag inside the Image to hide loading spinner when displayed again.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // console.log('modal-image onImageLoad', event);
        // console.log('modal-image onImageLoad images before', this.images);
        // sets as previously loaded the image with index specified by `event.status`
        this.images = this.images.map((/**
         * @param {?} img
         * @return {?}
         */
        function (img) {
            if (img && img.id === event.id) {
                return Object.assign({}, img, { previouslyLoaded: event.status });
            }
            return img;
        }));
        // console.log('modal-image onImageLoad images after', this.images);
    };
    /**
     * Method called when a dot is clicked and used to update the current image.
     * @param number index of the clicked dot
     */
    /**
     * Method called when a dot is clicked and used to update the current image.
     * @param {?} index
     * @return {?}
     */
    ModalGalleryComponent.prototype.onClickDot = /**
     * Method called when a dot is clicked and used to update the current image.
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.currentImage = this.images[index];
    };
    /**
     * Method called when an image preview is clicked and used to update the current image.
     * @param Image preview image
     */
    /**
     * Method called when an image preview is clicked and used to update the current image.
     * @param {?} event
     * @return {?}
     */
    ModalGalleryComponent.prototype.onClickPreview = /**
     * Method called when an image preview is clicked and used to update the current image.
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.onChangeCurrentImage(event);
    };
    // onClickArrow(event: InteractionEvent) {
    //   // TODO validate before to emit
    //   this.arrow.emit(event);
    // }
    /**
     * Method to download the current image, only if `downloadable` is true.
     * It contains also a logic to enable downloading features also for IE11.
     */
    // onClickArrow(event: InteractionEvent) {
    //   // TODO validate before to emit
    //   this.arrow.emit(event);
    // }
    /**
     * Method to download the current image, only if `downloadable` is true.
     * It contains also a logic to enable downloading features also for IE11.
     * @return {?}
     */
    ModalGalleryComponent.prototype.downloadImage = 
    // onClickArrow(event: InteractionEvent) {
    //   // TODO validate before to emit
    //   this.arrow.emit(event);
    // }
    /**
     * Method to download the current image, only if `downloadable` is true.
     * It contains also a logic to enable downloading features also for IE11.
     * @return {?}
     */
    function () {
        if (this.currentImageConfig && !this.currentImageConfig.downloadable) {
            return;
        }
        // If IE11 or Microsoft Edge use msSaveBlob(...)
        if (this.isIEorEdge()) {
            // I cannot use fetch API because IE11 doesn't support it,
            // so I have to switch to XMLHttpRequest
            this.downloadImageOnlyIEorEdge();
        }
        else {
            // for all other browsers
            this.downloadImageAllBrowsers();
        }
    };
    /**
     * Method to cleanup resources. In fact, this will reset keyboard's service.
     * This is an Angular's lifecycle hook that is called when this component is destroyed.
     */
    /**
     * Method to cleanup resources. In fact, this will reset keyboard's service.
     * This is an Angular's lifecycle hook that is called when this component is destroyed.
     * @return {?}
     */
    ModalGalleryComponent.prototype.ngOnDestroy = /**
     * Method to cleanup resources. In fact, this will reset keyboard's service.
     * This is an Angular's lifecycle hook that is called when this component is destroyed.
     * @return {?}
     */
    function () {
        this.keyboardService.reset();
        this.idValidatorService.remove(this.id);
        if (this.galleryServiceNavigateSubscription) {
            this.galleryServiceNavigateSubscription.unsubscribe();
        }
        if (this.galleryServiceCloseSubscription) {
            this.galleryServiceCloseSubscription.unsubscribe();
        }
        if (this.galleryServiceUpdateSubscription) {
            this.galleryServiceUpdateSubscription.unsubscribe();
        }
        if (this.galleryServiceAutoPlaySubscription) {
            this.galleryServiceAutoPlaySubscription.unsubscribe();
        }
    };
    /**
     * Private method to download the current image for all browsers except for IE11.
     */
    /**
     * Private method to download the current image for all browsers except for IE11.
     * @private
     * @return {?}
     */
    ModalGalleryComponent.prototype.downloadImageAllBrowsers = /**
     * Private method to download the current image for all browsers except for IE11.
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var link = document.createElement('a');
        link.href = (/** @type {?} */ (this.currentImage.modal.img));
        link.setAttribute('download', this.getFileName(this.currentImage));
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
    };
    /**
     * Private method to download the current image only for IE11 using
     * custom javascript's methods available only on IE.
     */
    /**
     * Private method to download the current image only for IE11 using
     * custom javascript's methods available only on IE.
     * @private
     * @return {?}
     */
    ModalGalleryComponent.prototype.downloadImageOnlyIEorEdge = /**
     * Private method to download the current image only for IE11 using
     * custom javascript's methods available only on IE.
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (isPlatformBrowser(this.platformId)) {
            /** @type {?} */
            var req_1 = new XMLHttpRequest();
            req_1.open('GET', (/** @type {?} */ (this.currentImage.modal.img)), true);
            req_1.responseType = 'arraybuffer';
            req_1.onload = (/**
             * @param {?} event
             * @return {?}
             */
            function (event) {
                /** @type {?} */
                var blob = new Blob([req_1.response], { type: 'image/png' });
                window.navigator.msSaveBlob(blob, _this.getFileName(_this.currentImage));
            });
            req_1.send();
        }
    };
    /**
     * Private method to get the `ButtonEvent` to emit, merging the input `ButtonEvent`
     * with the current image.
     * @param ButtonEvent event payload to return
     * @returns ButtonEvent event payload with the current image included
     */
    /**
     * Private method to get the `ButtonEvent` to emit, merging the input `ButtonEvent`
     * with the current image.
     * @private
     * @param {?} event
     * @return {?} ButtonEvent event payload with the current image included
     */
    ModalGalleryComponent.prototype.getButtonEventToEmit = /**
     * Private method to get the `ButtonEvent` to emit, merging the input `ButtonEvent`
     * with the current image.
     * @private
     * @param {?} event
     * @return {?} ButtonEvent event payload with the current image included
     */
    function (event) {
        return Object.assign(event, { image: this.currentImage });
    };
    /**
     * Private method to get the file name from an input path.
     * This is used either to get the image's name from its path or from the Image itself,
     * if specified as 'downloadFileName' by the user.
     * @param Image image to extract its file name
     * @returns string string file name of the input image.
     */
    /**
     * Private method to get the file name from an input path.
     * This is used either to get the image's name from its path or from the Image itself,
     * if specified as 'downloadFileName' by the user.
     * @private
     * @param {?} image
     * @return {?} string string file name of the input image.
     */
    ModalGalleryComponent.prototype.getFileName = /**
     * Private method to get the file name from an input path.
     * This is used either to get the image's name from its path or from the Image itself,
     * if specified as 'downloadFileName' by the user.
     * @private
     * @param {?} image
     * @return {?} string string file name of the input image.
     */
    function (image) {
        if (!image.modal.downloadFileName || image.modal.downloadFileName.length === 0) {
            return ((/** @type {?} */ (this.currentImage.modal.img))).replace(/^.*[\\\/]/, '');
        }
        else {
            return image.modal.downloadFileName;
        }
    };
    /**
     * Private method to initialize `images` as array of `Image`s.
     * Also, it will emit ImageowmodaModalEvent to say that images are loaded.
     */
    /**
     * Private method to initialize `images` as array of `Image`s.
     * Also, it will emit ImageowmodaModalEvent to say that images are loaded.
     * @private
     * @return {?}
     */
    ModalGalleryComponent.prototype.initImages = /**
     * Private method to initialize `images` as array of `Image`s.
     * Also, it will emit ImageowmodaModalEvent to say that images are loaded.
     * @private
     * @return {?}
     */
    function () {
        // I'm not cloning the array, but I'm doing this to cast it to an array of InternalLibImages
        this.images = (/** @type {?} */ (this.modalImages));
        this.hasData.emit(new ImageModalEvent(Action.LOAD, true));
        this.showGallery = this.images.length > 0;
    };
    /**
     * Private method to emit events when either the last or the first image are visible.
     * @param action Enum of type Action that represents the source of the event that changed the
     *  current image to the first one or the last one.
     * @param indexToCheck is the index number of the image (the first or the last one).
     */
    /**
     * Private method to emit events when either the last or the first image are visible.
     * @private
     * @param {?} action Enum of type Action that represents the source of the event that changed the
     *  current image to the first one or the last one.
     * @param {?} indexToCheck is the index number of the image (the first or the last one).
     * @return {?}
     */
    ModalGalleryComponent.prototype.emitBoundaryEvent = /**
     * Private method to emit events when either the last or the first image are visible.
     * @private
     * @param {?} action Enum of type Action that represents the source of the event that changed the
     *  current image to the first one or the last one.
     * @param {?} indexToCheck is the index number of the image (the first or the last one).
     * @return {?}
     */
    function (action, indexToCheck) {
        // to emit first/last event
        switch (indexToCheck) {
            case 0:
                this.firstImage.emit(new ImageModalEvent(action, true));
                break;
            case this.images.length - 1:
                this.lastImage.emit(new ImageModalEvent(action, true));
                break;
        }
    };
    /**
     * Private method to check if this library is running on
     * Microsoft browsers or not (i.e. it detects both IE11 and Edge)
     * supporting also Server-Side Rendering.
     * Inspired by https://msdn.microsoft.com/it-it/library/hh779016(v=vs.85).aspx
     * @returns any the result
     */
    /**
     * Private method to check if this library is running on
     * Microsoft browsers or not (i.e. it detects both IE11 and Edge)
     * supporting also Server-Side Rendering.
     * Inspired by https://msdn.microsoft.com/it-it/library/hh779016(v=vs.85).aspx
     * @private
     * @return {?} any the result
     */
    ModalGalleryComponent.prototype.isIEorEdge = /**
     * Private method to check if this library is running on
     * Microsoft browsers or not (i.e. it detects both IE11 and Edge)
     * supporting also Server-Side Rendering.
     * Inspired by https://msdn.microsoft.com/it-it/library/hh779016(v=vs.85).aspx
     * @private
     * @return {?} any the result
     */
    function () {
        if (isPlatformBrowser(this.platformId)) {
            // if both Blob constructor and msSaveOrOpenBlob are supported by the current browser
            return window.Blob && window.navigator.msSaveOrOpenBlob;
        }
        if (isPlatformServer(this.platformId)) {
            // server only
            return true;
        }
    };
    /** @nocollapse */
    ModalGalleryComponent.ctorParameters = function () { return [
        { type: KeyboardService },
        { type: GalleryService },
        { type: Object, decorators: [{ type: Inject, args: [PLATFORM_ID,] }] },
        { type: ChangeDetectorRef },
        { type: IdValidatorService }
    ]; };
    ModalGalleryComponent.propDecorators = {
        id: [{ type: Input }],
        modalImages: [{ type: Input }],
        buttonsConfig: [{ type: Input }],
        enableCloseOutside: [{ type: Input }],
        currentImageConfig: [{ type: Input }],
        dotsConfig: [{ type: Input }],
        previewConfig: [{ type: Input }],
        slideConfig: [{ type: Input }],
        accessibilityConfig: [{ type: Input }],
        keyboardConfig: [{ type: Input }],
        plainGalleryConfig: [{ type: Input }],
        close: [{ type: Output }],
        show: [{ type: Output }],
        firstImage: [{ type: Output }],
        lastImage: [{ type: Output }],
        hasData: [{ type: Output }],
        buttonBeforeHook: [{ type: Output }],
        buttonAfterHook: [{ type: Output }],
        arrow: [{ type: Output }],
        currentImageComponent: [{ type: ViewChild, args: [CurrentImageComponent,] }],
        onPopState: [{ type: HostListener, args: ['window:popstate', ['$event'],] }]
    };
ModalGalleryComponent.ɵfac = function ModalGalleryComponent_Factory(t) { return new (t || ModalGalleryComponent)(ɵngcc0.ɵɵdirectiveInject(ɵngcc1.KeyboardService), ɵngcc0.ɵɵdirectiveInject(ɵngcc2.GalleryService), ɵngcc0.ɵɵdirectiveInject(PLATFORM_ID), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef), ɵngcc0.ɵɵdirectiveInject(ɵngcc3.IdValidatorService)); };
ModalGalleryComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: ModalGalleryComponent, selectors: [["ks-modal-gallery"]], viewQuery: function ModalGalleryComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵviewQuery(CurrentImageComponent, true);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.currentImageComponent = _t.first);
    } }, hostBindings: function ModalGalleryComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("popstate", function ModalGalleryComponent_popstate_HostBindingHandler($event) { return ctx.onPopState($event); }, false, ɵngcc0.ɵɵresolveWindow);
    } }, inputs: { enableCloseOutside: "enableCloseOutside", accessibilityConfig: "accessibilityConfig", id: "id", modalImages: "modalImages", buttonsConfig: "buttonsConfig", currentImageConfig: "currentImageConfig", dotsConfig: "dotsConfig", previewConfig: "previewConfig", slideConfig: "slideConfig", keyboardConfig: "keyboardConfig", plainGalleryConfig: "plainGalleryConfig" }, outputs: { close: "close", show: "show", firstImage: "firstImage", lastImage: "lastImage", hasData: "hasData", buttonBeforeHook: "buttonBeforeHook", buttonAfterHook: "buttonAfterHook", arrow: "arrow" }, exportAs: ["ksModalGallery"], features: [ɵngcc0.ɵɵNgOnChangesFeature], decls: 3, vars: 7, consts: [[3, "images", "showGallery", "plainGalleryConfig", "accessibilityConfig", "show"], [3, "isOpen", "accessibilityConfig"], ["id", "modal-gallery-wrapper", "ksClickOutside", "", 3, "title", "clickOutsideEnable", "clickOutside", 4, "ngIf"], ["id", "modal-gallery-wrapper", "ksClickOutside", "", 3, "title", "clickOutsideEnable", "clickOutside"], ["id", "flex-min-height-ie-fix"], ["id", "modal-gallery-container"], [3, "currentImage", "buttonsConfig", "delete", "navigate", "download", "close", "fullscreen", "customEmit"], [3, "id", "images", "currentImage", "isOpen", "slideConfig", "keyboardConfig", "accessibilityConfig", "currentImageConfig", "loadImage", "changeImage", "close"], [3, "images", "currentImage", "dotsConfig", "accessibilityConfig", "clickDot"], [3, "images", "currentImage", "previewConfig", "accessibilityConfig", "clickPreview"]], template: function ModalGalleryComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "ks-plain-gallery", 0);
        ɵngcc0.ɵɵlistener("show", function ModalGalleryComponent_Template_ks_plain_gallery_show_0_listener($event) { return ctx.onShowModalGallery($event); });
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelement(1, "ks-background", 1);
        ɵngcc0.ɵɵtemplate(2, ModalGalleryComponent_div_2_Template, 8, 21, "div", 2);
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("images", ctx.images)("showGallery", ctx.showGallery && ctx.isPlainGalleryVisible())("plainGalleryConfig", ctx.plainGalleryConfig)("accessibilityConfig", ctx.accessibilityConfig);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("isOpen", ctx.opened)("accessibilityConfig", ctx.accessibilityConfig);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", ctx.opened);
    } }, directives: [ɵngcc4.PlainGalleryComponent, ɵngcc5.BackgroundComponent, ɵngcc6.NgIf, ɵngcc7.ClickOutsideDirective, ɵngcc8.UpperButtonsComponent, ɵngcc9.CurrentImageComponent, ɵngcc10.DotsComponent, ɵngcc11.PreviewsComponent], styles: ["#modal-gallery-wrapper[_ngcontent-%COMP%]{position:fixed;top:0;left:0;width:100vw;height:100vh;z-index:10000}#flex-min-height-ie-fix[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:center}#modal-gallery-container[_ngcontent-%COMP%]{display:flex;flex-direction:column;justify-content:space-between;min-height:100vh}"], changeDetection: 0 });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(ModalGalleryComponent, [{
        type: Component,
        args: [{
                selector: 'ks-modal-gallery',
                exportAs: 'ksModalGallery',
                template: "<ks-plain-gallery [images]=\"images\"\n                  [showGallery]=\"showGallery && isPlainGalleryVisible()\"\n                  [plainGalleryConfig]=\"plainGalleryConfig\"\n                  [accessibilityConfig]=\"accessibilityConfig\"\n                  (show)=\"onShowModalGallery($event)\"></ks-plain-gallery>\n\n<ks-background [isOpen]=\"opened\"\n               [accessibilityConfig]=\"accessibilityConfig\"></ks-background>\n\n<div id=\"modal-gallery-wrapper\" *ngIf=\"opened\"\n     [attr.aria-label]=\"accessibilityConfig.modalGalleryContentAriaLabel\"\n     [title]=\"accessibilityConfig.modalGalleryContentTitle\"\n     ksClickOutside [clickOutsideEnable]=\"enableCloseOutside\"\n     (clickOutside)=\"onClickOutside($event)\">\n\n  <div id=\"flex-min-height-ie-fix\">\n    <div id=\"modal-gallery-container\">\n\n      <ks-upper-buttons [currentImage]=\"currentImage\"\n                        [buttonsConfig]=\"buttonsConfig\"\n                        (delete)=\"onDelete($event)\"\n                        (navigate)=\"onNavigate($event)\"\n                        (download)=\"onDownload($event)\"\n                        (close)=\"onCloseGallery($event)\"\n                        (fullscreen)=\"onFullScreen($event)\"\n                        (customEmit)=\"onCustomEmit($event)\"></ks-upper-buttons>\n                        <!--(rotate)=\"onRotate($event)\"-->\n\n      <ks-current-image [id]=\"id\"\n                        [images]=\"images\"\n                        [currentImage]=\"currentImage\"\n                        [isOpen]=\"opened\"\n                        [slideConfig]=\"configSlide\"\n                        [keyboardConfig]=\"keyboardConfig\"\n                        [accessibilityConfig]=\"accessibilityConfig\"\n                        [currentImageConfig]=\"currentImageConfig\"\n                        (loadImage)=\"onImageLoad($event)\"\n                        (changeImage)=\"onChangeCurrentImage($event)\"\n                        (close)=\"onCloseGallery($event)\"></ks-current-image>\n\n      <div>\n        <ks-dots [images]=\"images\"\n                 [currentImage]=\"currentImage\"\n                 [dotsConfig]=\"dotsConfig\"\n                 [accessibilityConfig]=\"accessibilityConfig\"\n                 (clickDot)=\"onClickDot($event)\"></ks-dots>\n\n        <ks-previews [images]=\"images\"\n                     [currentImage]=\"currentImage\"\n                     [previewConfig]=\"previewConfig\"\n                     [accessibilityConfig]=\"accessibilityConfig\"\n                     (clickPreview)=\"onClickPreview($event)\"></ks-previews>\n      </div>\n    </div>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                styles: ["#modal-gallery-wrapper{position:fixed;top:0;left:0;width:100vw;height:100vh;z-index:10000}#flex-min-height-ie-fix{display:flex;flex-direction:column;justify-content:center}#modal-gallery-container{display:flex;flex-direction:column;justify-content:space-between;min-height:100vh}"]
            }]
    }], function () { return [{ type: ɵngcc1.KeyboardService }, { type: ɵngcc2.GalleryService }, { type: Object, decorators: [{
                type: Inject,
                args: [PLATFORM_ID]
            }] }, { type: ɵngcc0.ChangeDetectorRef }, { type: ɵngcc3.IdValidatorService }]; }, { enableCloseOutside: [{
            type: Input
        }], accessibilityConfig: [{
            type: Input
        }], close: [{
            type: Output
        }], show: [{
            type: Output
        }], firstImage: [{
            type: Output
        }], lastImage: [{
            type: Output
        }], hasData: [{
            type: Output
        }], buttonBeforeHook: [{
            type: Output
        }], buttonAfterHook: [{
            type: Output
        }], arrow: [{
            type: Output
        }], onPopState: [{
            type: HostListener,
            args: ['window:popstate', ['$event']]
        }], id: [{
            type: Input
        }], modalImages: [{
            type: Input
        }], buttonsConfig: [{
            type: Input
        }], currentImageConfig: [{
            type: Input
        }], dotsConfig: [{
            type: Input
        }], previewConfig: [{
            type: Input
        }], slideConfig: [{
            type: Input
        }], keyboardConfig: [{
            type: Input
        }], plainGalleryConfig: [{
            type: Input
        }], currentImageComponent: [{
            type: ViewChild,
            args: [CurrentImageComponent]
        }] }); })();
    return ModalGalleryComponent;
}());
export { ModalGalleryComponent };
if (false) {
    /**
     * Unique id (>=0) of the current instance of this library. This is useful when you are using
     * the service to call modal gallery without open it manually.
     * @type {?}
     */
    ModalGalleryComponent.prototype.id;
    /**
     * Array of `Image` that represent the model of this library with all images, thumbs and so on.
     * @type {?}
     */
    ModalGalleryComponent.prototype.modalImages;
    /**
     * Object of type `ButtonsConfig` to show/hide buttons.
     * @type {?}
     */
    ModalGalleryComponent.prototype.buttonsConfig;
    /**
     * Boolean to enable modal-gallery close behaviour when clicking
     * on the semi-transparent background. Enabled by default.
     * @type {?}
     */
    ModalGalleryComponent.prototype.enableCloseOutside;
    /**
     * Interface to configure current image in modal-gallery.
     * For instance you can disable navigation on click on current image (enabled by default).
     * @type {?}
     */
    ModalGalleryComponent.prototype.currentImageConfig;
    /**
     * Object of type `DotsConfig` to init DotsComponent's features.
     * For instance, it contains a param to show/hide dots.
     * @type {?}
     */
    ModalGalleryComponent.prototype.dotsConfig;
    /**
     * Object of type `PreviewConfig` to init PreviewsComponent's features.
     * For instance, it contains a param to show/hide previews.
     * @type {?}
     */
    ModalGalleryComponent.prototype.previewConfig;
    /**
     * Object of type `SlideConfig` to init side previews and `infinite sliding`.
     * @type {?}
     */
    ModalGalleryComponent.prototype.slideConfig;
    /**
     * Object of type `AccessibilityConfig` to init custom accessibility features.
     * For instance, it contains titles, alt texts, aria-labels and so on.
     * @type {?}
     */
    ModalGalleryComponent.prototype.accessibilityConfig;
    /**
     * Object of type `KeyboardConfig` to assign custom keys to ESC, RIGHT and LEFT keyboard's actions.
     * @type {?}
     */
    ModalGalleryComponent.prototype.keyboardConfig;
    /**
     * Object of type `PlainGalleryConfig` to configure the plain gallery.
     * @type {?}
     */
    ModalGalleryComponent.prototype.plainGalleryConfig;
    /**
     * Output to emit an event when the modal gallery is closed.
     * @type {?}
     */
    ModalGalleryComponent.prototype.close;
    /**
     * Output to emit an event when an image is changed.
     * @type {?}
     */
    ModalGalleryComponent.prototype.show;
    /**
     * Output to emit an event when the current image is the first one.
     * @type {?}
     */
    ModalGalleryComponent.prototype.firstImage;
    /**
     * Output to emit an event when the current image is the last one.
     * @type {?}
     */
    ModalGalleryComponent.prototype.lastImage;
    /**
     * Output to emit an event when the modal gallery is closed.
     * @type {?}
     */
    ModalGalleryComponent.prototype.hasData;
    /**
     * Output to emit an event when a button is clicked, but before that the action is triggered.
     * @type {?}
     */
    ModalGalleryComponent.prototype.buttonBeforeHook;
    /**
     * Output to emit an event when a button is clicked, but after that the action is triggered.
     * @type {?}
     */
    ModalGalleryComponent.prototype.buttonAfterHook;
    /**
     * Output to emit an event when someone clicks either an arrow of modal gallery or also in previews.
     * @type {?}
     */
    ModalGalleryComponent.prototype.arrow;
    /**
     * Reference to the CurrentImageComponent to invoke methods on it.
     * @type {?}
     */
    ModalGalleryComponent.prototype.currentImageComponent;
    /**
     * Boolean that it is true if the modal gallery is visible. False by default.
     * @type {?}
     */
    ModalGalleryComponent.prototype.opened;
    /**
     * Boolean to open the modal gallery. False by default.
     * @type {?}
     */
    ModalGalleryComponent.prototype.showGallery;
    /**
     * Array of `InternalLibImage` representing the model of this library with all images, thumbs and so on.
     * @type {?}
     */
    ModalGalleryComponent.prototype.images;
    /**
     * `Image` that is visible right now.
     * @type {?}
     */
    ModalGalleryComponent.prototype.currentImage;
    /**
     * Object of type `SlideConfig` passed to currentImage and used to handle play/stop features. This field is initialized
     * applying transformations, default values and so on to the input of the same type.
     * @type {?}
     */
    ModalGalleryComponent.prototype.configSlide;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.galleryServiceNavigateSubscription;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.galleryServiceCloseSubscription;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.galleryServiceUpdateSubscription;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.galleryServiceAutoPlaySubscription;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.keyboardService;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.galleryService;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.platformId;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.changeDetectorRef;
    /**
     * @type {?}
     * @private
     */
    ModalGalleryComponent.prototype.idValidatorService;
}
