// ngx-joyride/esm2015/lib/services/logger.service.js
import { Injectable } from '@angular/core';
import { JoyrideOptionsService } from './joyride-options.service';
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from './joyride-options.service';
const JOYRIDE = 'ngx-joyride:::';
export class LoggerService {
    constructor(optionService) {
        this.optionService = optionService;
    }
    debug(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.debug(JOYRIDE + message, data);
        }
    }
    info(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.info(JOYRIDE + message, data);
        }
    }
    warn(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.warn(JOYRIDE + message, data);
        }
    }
    error(message, data = "") {
        if (this.optionService.areLogsEnabled()) {
            console.error(JOYRIDE + message, data);
        }
    }
}
LoggerService.ɵfac = function LoggerService_Factory(t) { return new (t || LoggerService)(ɵngcc0.ɵɵinject(ɵngcc1.JoyrideOptionsService)); };
LoggerService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: LoggerService, factory: LoggerService.ɵfac });
LoggerService.ctorParameters = () => [
    { type: JoyrideOptionsService }
];
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(LoggerService, [{
        type: Injectable
    }], function () { return [{ type: ɵngcc1.JoyrideOptionsService }]; }, null); })();
